{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e04b6945-f40a-4298-9e3b-bf341c50d2e2",
   "metadata": {},
   "source": [
    "# Analyze targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9af008c9-d457-4675-8ea2-1ef1b4118c3f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "31263f08-dc84-4fa9-b970-45488e69d01a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/code/Isoprenol_CRISPRi/DavidCarruthers/make target list'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "332018c1-2d80-4fd6-8eeb-b61ba523d15c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_target = pd.read_csv('../DBTL0_results/df_targets_final.csv')\n",
    "df_isoprenol = pd.read_csv('../DBTL0_results/df_isoprenol_final.csv')\n",
    "target_id = pd.read_csv('../DBTL0_results/filtered_target_identity.csv')\n",
    "df2 = pd.read_csv('../DBTL0_results/df2_dbtl0_from_edd.csv', index_col = 0)\n",
    "df_target_nostats = df_target.iloc[:128,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f98fb453-68ed-4734-8084-054b14ef1e11",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Control-R1', 'Control-R10', 'Control-R11', 'Control-R12',\n",
       "       'Control-R13', 'Control-R14', 'Control-R15', 'Control-R16',\n",
       "       'Control-R17', 'Control-R18', 'Control-R2', 'Control-R3',\n",
       "       'Control-R4', 'Control-R5', 'Control-R6', 'Control-R7',\n",
       "       'Control-R8', 'Control-R9', 'PP_0001-R1', 'PP_0001-R2',\n",
       "       'PP_0001-R3', 'PP_0103-R1', 'PP_0103-R2', 'PP_0103-R3',\n",
       "       'PP_0104-R1', 'PP_0104-R2', 'PP_0104-R3', 'PP_0105-R1',\n",
       "       'PP_0105-R2', 'PP_0105-R3', 'PP_0106-R1', 'PP_0106-R2',\n",
       "       'PP_0106-R3', 'PP_0168-R1', 'PP_0168-R2', 'PP_0168-R3',\n",
       "       'PP_0225-R1', 'PP_0225-R2', 'PP_0225-R3', 'PP_0226-R1',\n",
       "       'PP_0226-R2', 'PP_0226-R3', 'PP_0277-R1', 'PP_0277-R2',\n",
       "       'PP_0277-R3', 'PP_0338-R1', 'PP_0338-R2', 'PP_0338-R3',\n",
       "       'PP_0362-R1', 'PP_0362-R2', 'PP_0362-R3', 'PP_0368-R1',\n",
       "       'PP_0368-R2', 'PP_0368-R3', 'PP_0437-R1', 'PP_0437-R2',\n",
       "       'PP_0437-R3', 'PP_0528-R1', 'PP_0528-R2', 'PP_0528-R3',\n",
       "       'PP_0548-R1', 'PP_0548-R2', 'PP_0548-R3', 'PP_0582-R1',\n",
       "       'PP_0582-R2', 'PP_0582-R3', 'PP_0597-R1', 'PP_0597-R2',\n",
       "       'PP_0597-R3', 'PP_0654-R1', 'PP_0654-R2', 'PP_0654-R3',\n",
       "       'PP_0658-R1', 'PP_0658-R2', 'PP_0658-R3', 'PP_0691-R1',\n",
       "       'PP_0691-R2', 'PP_0691-R3', 'PP_0751-R1', 'PP_0751-R2',\n",
       "       'PP_0751-R3', 'PP_0774-R1', 'PP_0774-R2', 'PP_0774-R3',\n",
       "       'PP_0806-R1', 'PP_0806-R2', 'PP_0806-R3', 'PP_0812-R1',\n",
       "       'PP_0812-R4', 'PP_0812-R2', 'PP_0812-R5', 'PP_0812-R3',\n",
       "       'PP_0812-R6', 'PP_0813-R1', 'PP_0813-R4', 'PP_0813-R2',\n",
       "       'PP_0813-R5', 'PP_0813-R3', 'PP_0813-R6', 'PP_0814-R1',\n",
       "       'PP_0814-R2', 'PP_0814-R3', 'PP_0815-R1', 'PP_0815-R2',\n",
       "       'PP_0815-R3', 'PP_0897-R1', 'PP_0897-R2', 'PP_0897-R3',\n",
       "       'PP_0944-R1', 'PP_0944-R2', 'PP_0944-R3', 'PP_0999-R1',\n",
       "       'PP_0999-R2', 'PP_0999-R3', 'PP_1023-R1', 'PP_1023-R2',\n",
       "       'PP_1023-R3', 'PP_1031-R1', 'PP_1031-R2', 'PP_1031-R3',\n",
       "       'PP_1157-R1', 'PP_1157-R2', 'PP_1157-R3', 'PP_1240-R1',\n",
       "       'PP_1240-R2', 'PP_1240-R3', 'PP_1251-R1', 'PP_1251-R2',\n",
       "       'PP_1251-R3', 'PP_1317-R1', 'PP_1317-R2', 'PP_1317-R3',\n",
       "       'PP_1318-R1', 'PP_1318-R2', 'PP_1318-R3', 'PP_1319-R1',\n",
       "       'PP_1319-R2', 'PP_1319-R3', 'PP_1394-R1', 'PP_1394-R2',\n",
       "       'PP_1394-R3', 'PP_1444-R1', 'PP_1444-R2', 'PP_1444-R3',\n",
       "       'PP_1457-R1', 'PP_1457-R2', 'PP_1457-R3', 'PP_1506-R1',\n",
       "       'PP_1506-R2', 'PP_1506-R3', 'PP_1607_NT1-R1', 'PP_1607_NT1-R2',\n",
       "       'PP_1607_NT1-R3', 'PP_1610-R1', 'PP_1610-R2', 'PP_1610-R3',\n",
       "       'PP_1620-R1', 'PP_1620-R2', 'PP_1620-R3', 'PP_1664-R1',\n",
       "       'PP_1664-R2', 'PP_1664-R3', 'PP_1755-R1', 'PP_1755-R2',\n",
       "       'PP_1755-R3', 'PP_1769-R1', 'PP_1769-R2', 'PP_1769-R3',\n",
       "       'PP_1770-R1', 'PP_1770-R2', 'PP_1770-R3', 'PP_1777-R1',\n",
       "       'PP_1777-R2', 'PP_1777-R3', 'PP_2095-R1', 'PP_2095-R2',\n",
       "       'PP_2095-R3', 'PP_2112-R1', 'PP_2112-R2', 'PP_2112-R3',\n",
       "       'PP_2136-R1', 'PP_2136-R2', 'PP_2136-R3', 'PP_2137-R1',\n",
       "       'PP_2137-R2', 'PP_2137-R3', 'PP_2213-R1', 'PP_2213-R2',\n",
       "       'PP_2213-R3', 'PP_2339-R1', 'PP_2339-R2', 'PP_2339-R3',\n",
       "       'PP_2471-R1', 'PP_2471-R2', 'PP_2471-R3', 'PP_2589-R1',\n",
       "       'PP_2589-R2', 'PP_2589-R3', 'PP_2925-R1', 'PP_2925-R2',\n",
       "       'PP_2925-R3', 'PP_3071-R1', 'PP_3071-R2', 'PP_3071-R3',\n",
       "       'PP_3122-R1', 'PP_3122-R2', 'PP_3122-R3', 'PP_3123-R1',\n",
       "       'PP_3123-R2', 'PP_3123-R3', 'PP_3280-R1', 'PP_3280-R2',\n",
       "       'PP_3280-R3', 'PP_3282-R1', 'PP_3282-R2', 'PP_3282-R3',\n",
       "       'PP_3355-R1', 'PP_3355-R2', 'PP_3355-R3', 'PP_3365-R1',\n",
       "       'PP_3365-R2', 'PP_3365-R3', 'PP_3394-R1', 'PP_3394-R2',\n",
       "       'PP_3394-R3', 'PP_3578-R1', 'PP_3578-R2', 'PP_3578-R3',\n",
       "       'PP_3744-R1', 'PP_3744-R2', 'PP_3744-R3', 'PP_3754-R1',\n",
       "       'PP_3754-R2', 'PP_3754-R3', 'PP_3755-R1', 'PP_3755-R2',\n",
       "       'PP_3755-R3', 'PP_4011-R1', 'PP_4011-R2', 'PP_4011-R3',\n",
       "       'PP_4012-R1', 'PP_4012-R2', 'PP_4012-R3', 'PP_4043-R1',\n",
       "       'PP_4043-R2', 'PP_4043-R3', 'PP_4116-R1', 'PP_4116-R2',\n",
       "       'PP_4116-R3', 'PP_4119-R1', 'PP_4119-R2', 'PP_4119-R3',\n",
       "       'PP_4120-R1', 'PP_4120-R2', 'PP_4120-R3', 'PP_4121-R1',\n",
       "       'PP_4121-R2', 'PP_4121-R3', 'PP_4122-R1', 'PP_4122-R2',\n",
       "       'PP_4122-R3', 'PP_4123-R1', 'PP_4123-R2', 'PP_4123-R3',\n",
       "       'PP_4128-R1', 'PP_4128-R2', 'PP_4128-R3', 'PP_4130-R1',\n",
       "       'PP_4130-R2', 'PP_4130-R3', 'PP_4159-R1', 'PP_4159-R2',\n",
       "       'PP_4159-R3', 'PP_4160-R1', 'PP_4160-R2', 'PP_4160-R3',\n",
       "       'PP_4161-R1', 'PP_4161-R2', 'PP_4161-R3', 'PP_4174-R1',\n",
       "       'PP_4174-R2', 'PP_4174-R3', 'PP_4175-R1', 'PP_4175-R2',\n",
       "       'PP_4175-R3', 'PP_4185-R1', 'PP_4185-R2', 'PP_4185-R3',\n",
       "       'PP_4186-R1', 'PP_4186-R2', 'PP_4186-R3', 'PP_4187-R1',\n",
       "       'PP_4187-R2', 'PP_4187-R3', 'PP_4188-R1', 'PP_4188-R2',\n",
       "       'PP_4188-R3', 'PP_4189-R1', 'PP_4189-R2', 'PP_4189-R3',\n",
       "       'PP_4191-R1', 'PP_4191-R2', 'PP_4191-R3', 'PP_4192-R1',\n",
       "       'PP_4192-R2', 'PP_4192-R3', 'PP_4193-R1', 'PP_4193-R2',\n",
       "       'PP_4193-R3', 'PP_4194_NT2-R1', 'PP_4194_NT2-R2', 'PP_4194_NT2-R3',\n",
       "       'PP_4266-R1', 'PP_4266-R2', 'PP_4266-R3', 'PP_4487-R1',\n",
       "       'PP_4487-R2', 'PP_4487-R3', 'PP_4549-R1', 'PP_4549-R2',\n",
       "       'PP_4549-R3', 'PP_4550-R1', 'PP_4550-R2', 'PP_4550-R3',\n",
       "       'PP_4635-R1', 'PP_4635-R2', 'PP_4635-R3', 'PP_4636-R1',\n",
       "       'PP_4636-R2', 'PP_4636-R3', 'PP_4650-R1', 'PP_4650-R2',\n",
       "       'PP_4650-R3', 'PP_4651-R1', 'PP_4651-R2', 'PP_4651-R3',\n",
       "       'PP_4667-R1', 'PP_4667-R2', 'PP_4667-R3', 'PP_4678-R1',\n",
       "       'PP_4678-R4', 'PP_4678-R2', 'PP_4678-R5', 'PP_4678-R3',\n",
       "       'PP_4678-R6', 'PP_4679-R1', 'PP_4679-R4', 'PP_4679-R2',\n",
       "       'PP_4679-R5', 'PP_4679-R3', 'PP_4679-R6', 'PP_4702-R1',\n",
       "       'PP_4702-R2', 'PP_4702-R3', 'PP_4737-R1', 'PP_4737-R2',\n",
       "       'PP_4737-R3', 'PP_4862-R1', 'PP_4862-R2', 'PP_4862-R3',\n",
       "       'PP_4894-R1', 'PP_4894-R2', 'PP_4894-R3', 'PP_5064-R1',\n",
       "       'PP_5064-R2', 'PP_5064-R3', 'PP_5128-R1', 'PP_5128-R2',\n",
       "       'PP_5128-R3', 'PP_5186-R1', 'PP_5186-R2', 'PP_5186-R3',\n",
       "       'PP_5203-R1', 'PP_5203-R2', 'PP_5203-R3', 'PP_5288-R1',\n",
       "       'PP_5288-R2', 'PP_5288-R3', 'PP_5335-R1', 'PP_5335-R2',\n",
       "       'PP_5335-R3', 'PP_5416-R1', 'PP_5416-R2', 'PP_5416-R3',\n",
       "       'PP_5417-R1', 'PP_5417-R2', 'PP_5417-R3', 'PP_5419-R1',\n",
       "       'PP_5419-R2', 'PP_5419-R3', 'PP_5420-R1', 'PP_5420-R2',\n",
       "       'PP_5420-R3'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2['Line Name'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b073be8d-ee75-4de2-ac35-d5e3e784b7ca",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "121"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_targets = [p for p in df_target['Target'].dropna().unique() if p[0:2] == 'PP']\n",
    "len(all_targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "239642b2-546f-4d20-ada9-f9d37c9f383a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_up = pd.read_excel('../putida_proteome_info/uniprotkb_edited.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "14d9ee14-f614-415d-8c59-d7b0d8b1ccb4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene_info_df = pd.read_excel('../putida_proteome_info/NamilPPDict.xlsx')\n",
    "gene_info_df['is_target'] = gene_info_df['Old_tag'].apply(lambda x: x in all_targets)\n",
    "df_info = gene_info_df[gene_info_df['is_target']].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0588b0cf-a523-4efd-983a-a284c090e4dd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>New_tag</th>\n",
       "      <th>Old_tag</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>Nickname</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Protein Length</th>\n",
       "      <th>Description</th>\n",
       "      <th>COG</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>is_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS00005</td>\n",
       "      <td>PP_0001</td>\n",
       "      <td>parB</td>\n",
       "      <td>parB</td>\n",
       "      <td>147</td>\n",
       "      <td>1019</td>\n",
       "      <td>-</td>\n",
       "      <td>290.0</td>\n",
       "      <td>Belongs to the ParB family</td>\n",
       "      <td>L</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K03497</td>\n",
       "      <td>-</td>\n",
       "      <td>P0A151</td>\n",
       "      <td>PARB_PSEPK</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Probable chromosome-partitioning protein ParB</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS00525</td>\n",
       "      <td>PP_0103</td>\n",
       "      <td>coxB</td>\n",
       "      <td>PP_0103</td>\n",
       "      <td>108810</td>\n",
       "      <td>109937</td>\n",
       "      <td>+</td>\n",
       "      <td>375.0</td>\n",
       "      <td>Subunits I and II form the functional core of ...</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>1.9.3.1</td>\n",
       "      <td>ko:K02275</td>\n",
       "      <td>R00081</td>\n",
       "      <td>Q88RM6</td>\n",
       "      <td>Q88RM6_PSEPK</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome c oxidase subunit 2 (EC 7.1.1.9)</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS00530</td>\n",
       "      <td>PP_0104</td>\n",
       "      <td>ctaD</td>\n",
       "      <td>ctaD</td>\n",
       "      <td>109981</td>\n",
       "      <td>111570</td>\n",
       "      <td>+</td>\n",
       "      <td>529.0</td>\n",
       "      <td>Cytochrome c oxidase is the component of the r...</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>1.10.3.10,1.9.3.1</td>\n",
       "      <td>ko:K02274,ko:K02298,ko:K15408</td>\n",
       "      <td>R00081,R11335</td>\n",
       "      <td>Q88RM5</td>\n",
       "      <td>Q88RM5_PSEPK</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome c oxidase subunit 1 (EC 7.1.1.9)</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS00535</td>\n",
       "      <td>PP_0105</td>\n",
       "      <td>ctaG</td>\n",
       "      <td>PP_0105</td>\n",
       "      <td>111580</td>\n",
       "      <td>112146</td>\n",
       "      <td>+</td>\n",
       "      <td>188.0</td>\n",
       "      <td>cytochrome c oxidase assembly protein</td>\n",
       "      <td>O</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02258</td>\n",
       "      <td>-</td>\n",
       "      <td>Q88RM4</td>\n",
       "      <td>Q88RM4_PSEPK</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome c oxidase assembly protein CtaG</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS00540</td>\n",
       "      <td>PP_0106</td>\n",
       "      <td>coxC</td>\n",
       "      <td>PP_0106</td>\n",
       "      <td>112163</td>\n",
       "      <td>113050</td>\n",
       "      <td>+</td>\n",
       "      <td>295.0</td>\n",
       "      <td>oxidase subunit</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>1.9.3.1</td>\n",
       "      <td>ko:K02276</td>\n",
       "      <td>R00081</td>\n",
       "      <td>Q88RM3</td>\n",
       "      <td>Q88RM3_PSEPK</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome aa3 subunit 3 (EC 7.1.1.9) (Cytochr...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS27785</td>\n",
       "      <td>PP_5335</td>\n",
       "      <td>purK</td>\n",
       "      <td>purK</td>\n",
       "      <td>6081951</td>\n",
       "      <td>6083033</td>\n",
       "      <td>-</td>\n",
       "      <td>360.0</td>\n",
       "      <td>Catalyzes the ATP-dependent conversion of 5- a...</td>\n",
       "      <td>F</td>\n",
       "      <td>GO:0000166,GO:0003674,GO:0003824,GO:0004638,GO...</td>\n",
       "      <td>6.3.4.18</td>\n",
       "      <td>ko:K01589</td>\n",
       "      <td>R07404</td>\n",
       "      <td>Q88C48</td>\n",
       "      <td>Q88C48_PSEPK</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>N5-carboxyaminoimidazole ribonucleotide syntha...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS28225</td>\n",
       "      <td>PP_5416</td>\n",
       "      <td>atpH</td>\n",
       "      <td>atpH</td>\n",
       "      <td>6179099</td>\n",
       "      <td>6179635</td>\n",
       "      <td>-</td>\n",
       "      <td>178.0</td>\n",
       "      <td>F(1)F(0) ATP synthase produces ATP from ADP in...</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0005215,GO:0005575,GO...</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02113</td>\n",
       "      <td>-</td>\n",
       "      <td>Q88BX1</td>\n",
       "      <td>ATPD_PSEPK</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ATP synthase subunit delta (ATP synthase F(1) ...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS28230</td>\n",
       "      <td>PP_5417</td>\n",
       "      <td>atpF</td>\n",
       "      <td>atpF</td>\n",
       "      <td>6179648</td>\n",
       "      <td>6180118</td>\n",
       "      <td>-</td>\n",
       "      <td>156.0</td>\n",
       "      <td>F(1)F(0) ATP synthase produces ATP from ADP in...</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0005215,GO:0005575,GO...</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02109</td>\n",
       "      <td>-</td>\n",
       "      <td>Q88BX0</td>\n",
       "      <td>ATPF_PSEPK</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ATP synthase subunit b (ATP synthase F(0) sect...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS28240</td>\n",
       "      <td>PP_5419</td>\n",
       "      <td>atpB</td>\n",
       "      <td>atpB</td>\n",
       "      <td>6180577</td>\n",
       "      <td>6181446</td>\n",
       "      <td>-</td>\n",
       "      <td>289.0</td>\n",
       "      <td>it plays a direct role in the translocation of...</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0005215,GO:0005575,GO...</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02108</td>\n",
       "      <td>-</td>\n",
       "      <td>Q88BW8</td>\n",
       "      <td>ATP6_PSEPK</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ATP synthase subunit a (ATP synthase F0 sector...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS28245</td>\n",
       "      <td>PP_5420</td>\n",
       "      <td>atpI</td>\n",
       "      <td>atpI</td>\n",
       "      <td>6181463</td>\n",
       "      <td>6181870</td>\n",
       "      <td>-</td>\n",
       "      <td>135.0</td>\n",
       "      <td>ATP synthase</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0005215,GO:0006139,GO...</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02116</td>\n",
       "      <td>-</td>\n",
       "      <td>P0A103</td>\n",
       "      <td>ATPZ_PSEPK</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>ATP synthase protein I</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Type     New_tag  Old_tag Preferred_name Nickname    Start      End  \\\n",
       "0    CDS  PP_RS00005  PP_0001           parB     parB      147     1019   \n",
       "1    CDS  PP_RS00525  PP_0103           coxB  PP_0103   108810   109937   \n",
       "2    CDS  PP_RS00530  PP_0104           ctaD     ctaD   109981   111570   \n",
       "3    CDS  PP_RS00535  PP_0105           ctaG  PP_0105   111580   112146   \n",
       "4    CDS  PP_RS00540  PP_0106           coxC  PP_0106   112163   113050   \n",
       "..   ...         ...      ...            ...      ...      ...      ...   \n",
       "116  CDS  PP_RS27785  PP_5335           purK     purK  6081951  6083033   \n",
       "117  CDS  PP_RS28225  PP_5416           atpH    atpH   6179099  6179635   \n",
       "118  CDS  PP_RS28230  PP_5417           atpF     atpF  6179648  6180118   \n",
       "119  CDS  PP_RS28240  PP_5419           atpB     atpB  6180577  6181446   \n",
       "120  CDS  PP_RS28245  PP_5420           atpI     atpI  6181463  6181870   \n",
       "\n",
       "    Strand  Protein Length                                        Description  \\\n",
       "0        -           290.0                         Belongs to the ParB family   \n",
       "1        +           375.0  Subunits I and II form the functional core of ...   \n",
       "2        +           529.0  Cytochrome c oxidase is the component of the r...   \n",
       "3        +           188.0              cytochrome c oxidase assembly protein   \n",
       "4        +           295.0                                    oxidase subunit   \n",
       "..     ...             ...                                                ...   \n",
       "116      -           360.0  Catalyzes the ATP-dependent conversion of 5- a...   \n",
       "117      -           178.0  F(1)F(0) ATP synthase produces ATP from ADP in...   \n",
       "118      -           156.0  F(1)F(0) ATP synthase produces ATP from ADP in...   \n",
       "119      -           289.0  it plays a direct role in the translocation of...   \n",
       "120      -           135.0                                       ATP synthase   \n",
       "\n",
       "    COG                                                GOs                 EC  \\\n",
       "0     L                                                  -                  -   \n",
       "1     C                                                  -            1.9.3.1   \n",
       "2     C                                                  -  1.10.3.10,1.9.3.1   \n",
       "3     O                                                  -                  -   \n",
       "4     C                                                  -            1.9.3.1   \n",
       "..   ..                                                ...                ...   \n",
       "116   F  GO:0000166,GO:0003674,GO:0003824,GO:0004638,GO...           6.3.4.18   \n",
       "117   C  GO:0003674,GO:0003824,GO:0005215,GO:0005575,GO...                  -   \n",
       "118   C  GO:0003674,GO:0003824,GO:0005215,GO:0005575,GO...                  -   \n",
       "119   C  GO:0003674,GO:0003824,GO:0005215,GO:0005575,GO...                  -   \n",
       "120   C  GO:0003674,GO:0003824,GO:0005215,GO:0006139,GO...                  -   \n",
       "\n",
       "                           KEGG_ko  KEGG_Reaction   Entry    Entry name  \\\n",
       "0                        ko:K03497              -  P0A151    PARB_PSEPK   \n",
       "1                        ko:K02275         R00081  Q88RM6  Q88RM6_PSEPK   \n",
       "2    ko:K02274,ko:K02298,ko:K15408  R00081,R11335  Q88RM5  Q88RM5_PSEPK   \n",
       "3                        ko:K02258              -  Q88RM4  Q88RM4_PSEPK   \n",
       "4                        ko:K02276         R00081  Q88RM3  Q88RM3_PSEPK   \n",
       "..                             ...            ...     ...           ...   \n",
       "116                      ko:K01589         R07404  Q88C48  Q88C48_PSEPK   \n",
       "117                      ko:K02113              -  Q88BX1    ATPD_PSEPK   \n",
       "118                      ko:K02109              -  Q88BX0    ATPF_PSEPK   \n",
       "119                      ko:K02108              -  Q88BW8    ATP6_PSEPK   \n",
       "120                      ko:K02116              -  P0A103    ATPZ_PSEPK   \n",
       "\n",
       "         Status                                       Protein name  is_target  \n",
       "0      reviewed      Probable chromosome-partitioning protein ParB       True  \n",
       "1    unreviewed        Cytochrome c oxidase subunit 2 (EC 7.1.1.9)       True  \n",
       "2    unreviewed        Cytochrome c oxidase subunit 1 (EC 7.1.1.9)       True  \n",
       "3    unreviewed         Cytochrome c oxidase assembly protein CtaG       True  \n",
       "4    unreviewed  Cytochrome aa3 subunit 3 (EC 7.1.1.9) (Cytochr...       True  \n",
       "..          ...                                                ...        ...  \n",
       "116  unreviewed  N5-carboxyaminoimidazole ribonucleotide syntha...       True  \n",
       "117    reviewed  ATP synthase subunit delta (ATP synthase F(1) ...       True  \n",
       "118    reviewed  ATP synthase subunit b (ATP synthase F(0) sect...       True  \n",
       "119    reviewed  ATP synthase subunit a (ATP synthase F0 sector...       True  \n",
       "120    reviewed                             ATP synthase protein I       True  \n",
       "\n",
       "[121 rows x 20 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8101d180-431f-4f76-9667-9549e4596ddd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_info['Old_tag'].to_csv('target_info_output/target_list.csv', header = False, index = False)\n",
    "df_info['Preferred_name'].to_csv('target_info_output/target_list_pref_name.csv', header = False, index = False)\n",
    "df_info['Entry'].to_csv('target_info_output/target_list_entry.csv', header = False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5b4be54c-4d9a-434a-bdec-b4b726ec3a0e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "62"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_info['pref_first3'] = [s[:3] for s in df_info['Preferred_name']]\n",
    "len(df_info['pref_first3'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fc7357b4-82a4-46c5-95e4-f4926e7c6fab",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>New_tag</th>\n",
       "      <th>Old_tag</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>Nickname</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Protein Length</th>\n",
       "      <th>Description</th>\n",
       "      <th>...</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>is_target</th>\n",
       "      <th>pref_first3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS00005</td>\n",
       "      <td>PP_0001</td>\n",
       "      <td>parB</td>\n",
       "      <td>parB</td>\n",
       "      <td>147</td>\n",
       "      <td>1019</td>\n",
       "      <td>-</td>\n",
       "      <td>290.0</td>\n",
       "      <td>Belongs to the ParB family</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K03497</td>\n",
       "      <td>-</td>\n",
       "      <td>P0A151</td>\n",
       "      <td>PARB_PSEPK</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Probable chromosome-partitioning protein ParB</td>\n",
       "      <td>True</td>\n",
       "      <td>par</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS00525</td>\n",
       "      <td>PP_0103</td>\n",
       "      <td>coxB</td>\n",
       "      <td>PP_0103</td>\n",
       "      <td>108810</td>\n",
       "      <td>109937</td>\n",
       "      <td>+</td>\n",
       "      <td>375.0</td>\n",
       "      <td>Subunits I and II form the functional core of ...</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>1.9.3.1</td>\n",
       "      <td>ko:K02275</td>\n",
       "      <td>R00081</td>\n",
       "      <td>Q88RM6</td>\n",
       "      <td>Q88RM6_PSEPK</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome c oxidase subunit 2 (EC 7.1.1.9)</td>\n",
       "      <td>True</td>\n",
       "      <td>cox</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDS</td>\n",
       "      <td>PP_RS00530</td>\n",
       "      <td>PP_0104</td>\n",
       "      <td>ctaD</td>\n",
       "      <td>ctaD</td>\n",
       "      <td>109981</td>\n",
       "      <td>111570</td>\n",
       "      <td>+</td>\n",
       "      <td>529.0</td>\n",
       "      <td>Cytochrome c oxidase is the component of the r...</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>1.10.3.10,1.9.3.1</td>\n",
       "      <td>ko:K02274,ko:K02298,ko:K15408</td>\n",
       "      <td>R00081,R11335</td>\n",
       "      <td>Q88RM5</td>\n",
       "      <td>Q88RM5_PSEPK</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome c oxidase subunit 1 (EC 7.1.1.9)</td>\n",
       "      <td>True</td>\n",
       "      <td>cta</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Type     New_tag  Old_tag Preferred_name Nickname   Start     End Strand  \\\n",
       "0  CDS  PP_RS00005  PP_0001           parB     parB     147    1019      -   \n",
       "1  CDS  PP_RS00525  PP_0103           coxB  PP_0103  108810  109937      +   \n",
       "2  CDS  PP_RS00530  PP_0104           ctaD     ctaD  109981  111570      +   \n",
       "\n",
       "   Protein Length                                        Description  ... GOs  \\\n",
       "0           290.0                         Belongs to the ParB family  ...   -   \n",
       "1           375.0  Subunits I and II form the functional core of ...  ...   -   \n",
       "2           529.0  Cytochrome c oxidase is the component of the r...  ...   -   \n",
       "\n",
       "                  EC                        KEGG_ko  KEGG_Reaction   Entry  \\\n",
       "0                  -                      ko:K03497              -  P0A151   \n",
       "1            1.9.3.1                      ko:K02275         R00081  Q88RM6   \n",
       "2  1.10.3.10,1.9.3.1  ko:K02274,ko:K02298,ko:K15408  R00081,R11335  Q88RM5   \n",
       "\n",
       "     Entry name      Status                                   Protein name  \\\n",
       "0    PARB_PSEPK    reviewed  Probable chromosome-partitioning protein ParB   \n",
       "1  Q88RM6_PSEPK  unreviewed    Cytochrome c oxidase subunit 2 (EC 7.1.1.9)   \n",
       "2  Q88RM5_PSEPK  unreviewed    Cytochrome c oxidase subunit 1 (EC 7.1.1.9)   \n",
       "\n",
       "  is_target  pref_first3  \n",
       "0      True          par  \n",
       "1      True          cox  \n",
       "2      True          cta  \n",
       "\n",
       "[3 rows x 21 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_info.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0490b665-749a-4192-a7b8-a4d2c762f16d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Gene Ontology (molecular function)</th>\n",
       "      <th>Gene Ontology (GO)</th>\n",
       "      <th>Gene Ontology IDs</th>\n",
       "      <th>Gene Ontology (cellular component)</th>\n",
       "      <th>Gene Ontology (biological process)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q88CC1</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>HGLS_PSEPK</td>\n",
       "      <td>2-oxoadipate dioxygenase/decarboxylase (EC 1.1...</td>\n",
       "      <td>hglS ydcJ PP_5260</td>\n",
       "      <td>Pseudomonas putida (strain ATCC 47054 / DSM 61...</td>\n",
       "      <td>464</td>\n",
       "      <td>dioxygenase activity [GO:0051213]</td>\n",
       "      <td>dioxygenase activity [GO:0051213]</td>\n",
       "      <td>GO:0051213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q88E10</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>MCPS_PSEPK</td>\n",
       "      <td>Methyl-accepting chemotaxis protein McpS</td>\n",
       "      <td>mcpS PP_4658</td>\n",
       "      <td>Pseudomonas putida (strain ATCC 47054 / DSM 61...</td>\n",
       "      <td>639</td>\n",
       "      <td>transmembrane signaling receptor activity [GO:...</td>\n",
       "      <td>plasma membrane [GO:0005886]; transmembrane si...</td>\n",
       "      <td>GO:0004888; GO:0005886; GO:0006935; GO:0007165</td>\n",
       "      <td>plasma membrane [GO:0005886]</td>\n",
       "      <td>chemotaxis [GO:0006935]; signal transduction [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q88E47</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>HGD_PSEPK</td>\n",
       "      <td>Homogentisate 1,2-dioxygenase (HGDO) (EC 1.13....</td>\n",
       "      <td>hmgA PP_4621</td>\n",
       "      <td>Pseudomonas putida (strain ATCC 47054 / DSM 61...</td>\n",
       "      <td>433</td>\n",
       "      <td>DNA-binding transcription repressor activity [...</td>\n",
       "      <td>protein-DNA complex [GO:0032993]; DNA-binding ...</td>\n",
       "      <td>GO:0000976; GO:0001217; GO:0004411; GO:0005506...</td>\n",
       "      <td>protein-DNA complex [GO:0032993]</td>\n",
       "      <td>L-phenylalanine catabolic process [GO:0006559]...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entry  Reviewed  Entry Name  \\\n",
       "0  Q88CC1  reviewed  HGLS_PSEPK   \n",
       "1  Q88E10  reviewed  MCPS_PSEPK   \n",
       "2  Q88E47  reviewed   HGD_PSEPK   \n",
       "\n",
       "                                       Protein names         Gene Names  \\\n",
       "0  2-oxoadipate dioxygenase/decarboxylase (EC 1.1...  hglS ydcJ PP_5260   \n",
       "1           Methyl-accepting chemotaxis protein McpS       mcpS PP_4658   \n",
       "2  Homogentisate 1,2-dioxygenase (HGDO) (EC 1.13....       hmgA PP_4621   \n",
       "\n",
       "                                            Organism  Length  \\\n",
       "0  Pseudomonas putida (strain ATCC 47054 / DSM 61...     464   \n",
       "1  Pseudomonas putida (strain ATCC 47054 / DSM 61...     639   \n",
       "2  Pseudomonas putida (strain ATCC 47054 / DSM 61...     433   \n",
       "\n",
       "                  Gene Ontology (molecular function)  \\\n",
       "0                  dioxygenase activity [GO:0051213]   \n",
       "1  transmembrane signaling receptor activity [GO:...   \n",
       "2  DNA-binding transcription repressor activity [...   \n",
       "\n",
       "                                  Gene Ontology (GO)  \\\n",
       "0                  dioxygenase activity [GO:0051213]   \n",
       "1  plasma membrane [GO:0005886]; transmembrane si...   \n",
       "2  protein-DNA complex [GO:0032993]; DNA-binding ...   \n",
       "\n",
       "                                   Gene Ontology IDs  \\\n",
       "0                                         GO:0051213   \n",
       "1     GO:0004888; GO:0005886; GO:0006935; GO:0007165   \n",
       "2  GO:0000976; GO:0001217; GO:0004411; GO:0005506...   \n",
       "\n",
       "  Gene Ontology (cellular component)  \\\n",
       "0                                NaN   \n",
       "1       plasma membrane [GO:0005886]   \n",
       "2   protein-DNA complex [GO:0032993]   \n",
       "\n",
       "                  Gene Ontology (biological process)  \n",
       "0                                                NaN  \n",
       "1  chemotaxis [GO:0006935]; signal transduction [...  \n",
       "2  L-phenylalanine catabolic process [GO:0006559]...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_up.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c76fc72e-93e4-4201-8d71-dd9003eeab43",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'abcd'.find('c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "73ea447b-55c4-4717-9e4d-696528f9c647",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def return_pp(x):\n",
    "    pp_idx = x.find('PP_')\n",
    "    if pp_idx != -1:\n",
    "        return x[pp_idx:pp_idx+7]\n",
    "    else:\n",
    "        return None\n",
    "df_up['pp'] = df_up['Gene Names'].apply(return_pp)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ce8449f4-e8e9-423d-a64a-12f683dcaead",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_up['pp'].unique()\n",
    "df_up['is_target'] = df_up['pp'].apply(lambda x: x in all_targets)\n",
    "df_up = df_up[df_up['is_target']].reset_index(drop = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c8851574-ff9d-453e-857e-215c8dda7e94",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df_merge shape: (120, 34)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pp</th>\n",
       "      <th>Entry_x</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Gene Ontology (molecular function)</th>\n",
       "      <th>Gene Ontology (GO)</th>\n",
       "      <th>...</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>Entry_y</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>is_target_y</th>\n",
       "      <th>pref_first3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP_2136</td>\n",
       "      <td>Q88L02</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>FADB_PSEPK</td>\n",
       "      <td>Fatty acid oxidation complex subunit alpha [In...</td>\n",
       "      <td>fadB PP_2136</td>\n",
       "      <td>Pseudomonas putida (strain ATCC 47054 / DSM 61...</td>\n",
       "      <td>715</td>\n",
       "      <td>3-hydroxyacyl-CoA dehydrogenase activity [GO:0...</td>\n",
       "      <td>fatty acid beta-oxidation multienzyme complex ...</td>\n",
       "      <td>...</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0003857,GO:0004165,GO...</td>\n",
       "      <td>1.1.1.35,4.2.1.17,5.1.2.3,5.3.3.8</td>\n",
       "      <td>ko:K01782,ko:K01825</td>\n",
       "      <td>R01975,R03026,R03045,R03276,R04137,R04170,R042...</td>\n",
       "      <td>Q88L02</td>\n",
       "      <td>FADB_PSEPK</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Fatty acid oxidation complex subunit alpha [In...</td>\n",
       "      <td>True</td>\n",
       "      <td>fad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP_1610</td>\n",
       "      <td>Q88MG1</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>PYRG_PSEPK</td>\n",
       "      <td>CTP synthase (EC 6.3.4.2) (Cytidine 5'-triphos...</td>\n",
       "      <td>pyrG PP_1610</td>\n",
       "      <td>Pseudomonas putida (strain ATCC 47054 / DSM 61...</td>\n",
       "      <td>542</td>\n",
       "      <td>ATP binding [GO:0005524]; CTP synthase activit...</td>\n",
       "      <td>ATP binding [GO:0005524]; CTP synthase activit...</td>\n",
       "      <td>...</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0003883,GO:0005488,GO...</td>\n",
       "      <td>6.3.4.2</td>\n",
       "      <td>ko:K01937</td>\n",
       "      <td>R00571,R00573</td>\n",
       "      <td>Q88MG1</td>\n",
       "      <td>PYRG_PSEPK</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>CTP synthase (EC 6.3.4.2) (Cytidine 5'-triphos...</td>\n",
       "      <td>True</td>\n",
       "      <td>pyr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP_4160</td>\n",
       "      <td>A0A140FWL1</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>A0A140FWL1_PSEPK</td>\n",
       "      <td>Potassium-transporting ATPase ATP-binding subu...</td>\n",
       "      <td>kdpB PP_4160</td>\n",
       "      <td>Pseudomonas putida (strain ATCC 47054 / DSM 61...</td>\n",
       "      <td>684</td>\n",
       "      <td>ATP binding [GO:0005524]; ATP hydrolysis activ...</td>\n",
       "      <td>plasma membrane [GO:0005886]; ATP binding [GO:...</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>3.6.3.12</td>\n",
       "      <td>ko:K01547</td>\n",
       "      <td>-</td>\n",
       "      <td>A0A140FWL1</td>\n",
       "      <td>A0A140FWL1_PSEPK</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Potassium-transporting ATPase ATP-binding subu...</td>\n",
       "      <td>True</td>\n",
       "      <td>kdp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        pp     Entry_x    Reviewed        Entry Name  \\\n",
       "0  PP_2136      Q88L02    reviewed        FADB_PSEPK   \n",
       "1  PP_1610      Q88MG1    reviewed        PYRG_PSEPK   \n",
       "2  PP_4160  A0A140FWL1  unreviewed  A0A140FWL1_PSEPK   \n",
       "\n",
       "                                       Protein names    Gene Names  \\\n",
       "0  Fatty acid oxidation complex subunit alpha [In...  fadB PP_2136   \n",
       "1  CTP synthase (EC 6.3.4.2) (Cytidine 5'-triphos...  pyrG PP_1610   \n",
       "2  Potassium-transporting ATPase ATP-binding subu...  kdpB PP_4160   \n",
       "\n",
       "                                            Organism  Length  \\\n",
       "0  Pseudomonas putida (strain ATCC 47054 / DSM 61...     715   \n",
       "1  Pseudomonas putida (strain ATCC 47054 / DSM 61...     542   \n",
       "2  Pseudomonas putida (strain ATCC 47054 / DSM 61...     684   \n",
       "\n",
       "                  Gene Ontology (molecular function)  \\\n",
       "0  3-hydroxyacyl-CoA dehydrogenase activity [GO:0...   \n",
       "1  ATP binding [GO:0005524]; CTP synthase activit...   \n",
       "2  ATP binding [GO:0005524]; ATP hydrolysis activ...   \n",
       "\n",
       "                                  Gene Ontology (GO)  ...  \\\n",
       "0  fatty acid beta-oxidation multienzyme complex ...  ...   \n",
       "1  ATP binding [GO:0005524]; CTP synthase activit...  ...   \n",
       "2  plasma membrane [GO:0005886]; ATP binding [GO:...  ...   \n",
       "\n",
       "                                                 GOs  \\\n",
       "0  GO:0003674,GO:0003824,GO:0003857,GO:0004165,GO...   \n",
       "1  GO:0003674,GO:0003824,GO:0003883,GO:0005488,GO...   \n",
       "2                                                  -   \n",
       "\n",
       "                                  EC              KEGG_ko  \\\n",
       "0  1.1.1.35,4.2.1.17,5.1.2.3,5.3.3.8  ko:K01782,ko:K01825   \n",
       "1                            6.3.4.2            ko:K01937   \n",
       "2                           3.6.3.12            ko:K01547   \n",
       "\n",
       "                                       KEGG_Reaction     Entry_y  \\\n",
       "0  R01975,R03026,R03045,R03276,R04137,R04170,R042...      Q88L02   \n",
       "1                                      R00571,R00573      Q88MG1   \n",
       "2                                                  -  A0A140FWL1   \n",
       "\n",
       "         Entry name      Status  \\\n",
       "0        FADB_PSEPK    reviewed   \n",
       "1        PYRG_PSEPK    reviewed   \n",
       "2  A0A140FWL1_PSEPK  unreviewed   \n",
       "\n",
       "                                        Protein name  is_target_y  pref_first3  \n",
       "0  Fatty acid oxidation complex subunit alpha [In...         True          fad  \n",
       "1  CTP synthase (EC 6.3.4.2) (Cytidine 5'-triphos...         True          pyr  \n",
       "2  Potassium-transporting ATPase ATP-binding subu...         True          kdp  \n",
       "\n",
       "[3 rows x 34 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_up_pp = df_up.copy().set_index('pp')\n",
    "df_info_pp = df_info.copy().rename(columns = {'Old_tag': 'pp'}).set_index('pp')\n",
    "df_merge = pd.merge(\n",
    "    left = df_up_pp,\n",
    "    right = df_info_pp,\n",
    "    how = 'inner',\n",
    "    left_index = True,\n",
    "    right_index = True\n",
    ").reset_index(drop = False)\n",
    "print(f'df_merge shape: {df_merge.shape}')\n",
    "df_merge.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae682f1b-7b29-4301-8e3c-338e983c09ce",
   "metadata": {},
   "source": [
    "Both datasets have an 'Entry' column - check that they match."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "889dfeb7-79ac-4731-ae6e-9bf1cb8c4c4c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are N = 1 targets where the 'Entry' columns do not match\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pp</th>\n",
       "      <th>Entry_x</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Gene Ontology (molecular function)</th>\n",
       "      <th>Gene Ontology (GO)</th>\n",
       "      <th>...</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>Entry_y</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>is_target_y</th>\n",
       "      <th>pref_first3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>PP_1157</td>\n",
       "      <td>Q877U6</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Q877U6_PSEPK</td>\n",
       "      <td>Acetolactate synthase</td>\n",
       "      <td>PP_1157 PP_3365</td>\n",
       "      <td>Pseudomonas putida (strain ATCC 47054 / DSM 61...</td>\n",
       "      <td>547</td>\n",
       "      <td>catalytic activity [GO:0003824]; magnesium ion...</td>\n",
       "      <td>catalytic activity [GO:0003824]; magnesium ion...</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>2.2.1.6</td>\n",
       "      <td>ko:K01652</td>\n",
       "      <td>R00006,R00014,R00226,R03050,R04672,R04673,R08648</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>bud</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         pp Entry_x    Reviewed    Entry Name          Protein names  \\\n",
       "90  PP_1157  Q877U6  unreviewed  Q877U6_PSEPK  Acetolactate synthase   \n",
       "\n",
       "         Gene Names                                           Organism  \\\n",
       "90  PP_1157 PP_3365  Pseudomonas putida (strain ATCC 47054 / DSM 61...   \n",
       "\n",
       "    Length                 Gene Ontology (molecular function)  \\\n",
       "90     547  catalytic activity [GO:0003824]; magnesium ion...   \n",
       "\n",
       "                                   Gene Ontology (GO)  ... GOs       EC  \\\n",
       "90  catalytic activity [GO:0003824]; magnesium ion...  ...   -  2.2.1.6   \n",
       "\n",
       "      KEGG_ko                                     KEGG_Reaction Entry_y  \\\n",
       "90  ko:K01652  R00006,R00014,R00226,R03050,R04672,R04673,R08648       -   \n",
       "\n",
       "   Entry name      Status Protein name  is_target_y  pref_first3  \n",
       "90          -  unreviewed            -         True          bud  \n",
       "\n",
       "[1 rows x 34 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"There are N = {sum(df_merge['Entry_x'] != df_merge['Entry_y'])} targets where the \\'Entry\\' columns do not match\")\n",
    "df_merge[df_merge['Entry_x'] != df_merge['Entry_y']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d90aecf3-f2dd-41ae-b091-4b337685dbc9",
   "metadata": {},
   "source": [
    "Save merged dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "84397480-e61e-4c37-aa48-f13e4d90381a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['pp', 'Entry_x', 'Reviewed', 'Entry Name', 'Protein names',\n",
       "       'Gene Names', 'Organism', 'Length',\n",
       "       'Gene Ontology (molecular function)', 'Gene Ontology (GO)',\n",
       "       'Gene Ontology IDs', 'Gene Ontology (cellular component)',\n",
       "       'Gene Ontology (biological process)', 'is_target_x', 'Type', 'New_tag',\n",
       "       'Preferred_name', 'Nickname', 'Start', 'End', 'Strand',\n",
       "       'Protein Length', 'Description', 'COG', 'GOs', 'EC', 'KEGG_ko',\n",
       "       'KEGG_Reaction', 'Entry_y', 'Entry name', 'Status', 'Protein name',\n",
       "       'is_target_y', 'pref_first3'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ffffe45a-b6ec-4857-a629-346a5bcd1f0b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_merge = df_merge.drop(columns=['Entry_y', 'Entry name', 'Status', 'Protein name', 'is_target_y', 'is_target_x', 'Start', 'End', 'Strand', 'Protein Length', 'Organism', 'Length', 'Type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "68e2ab34-d24d-41b7-b5a4-14e1cce860d8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_merge = df_merge.sort_values('pp').reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09fb61a2-e104-4e8f-a796-a1a0b5888fd1",
   "metadata": {},
   "source": [
    "## Incorporate information from GSM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e41c2687-807d-4536-b3e9-2385500f3ec3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting cobra\n",
      "  Downloading cobra-0.29.1-py2.py3-none-any.whl.metadata (9.3 kB)\n",
      "Collecting appdirs~=1.4 (from cobra)\n",
      "  Downloading appdirs-1.4.4-py2.py3-none-any.whl.metadata (9.0 kB)\n",
      "Collecting depinfo~=2.2 (from cobra)\n",
      "  Downloading depinfo-2.2.0-py3-none-any.whl.metadata (3.8 kB)\n",
      "Collecting diskcache~=5.0 (from cobra)\n",
      "  Downloading diskcache-5.6.3-py3-none-any.whl.metadata (20 kB)\n",
      "Collecting future (from cobra)\n",
      "  Downloading future-1.0.0-py3-none-any.whl.metadata (4.0 kB)\n",
      "Requirement already satisfied: httpx~=0.24 in /usr/local/art/.venv/lib/python3.12/site-packages (from cobra) (0.27.2)\n",
      "Collecting importlib-resources (from cobra)\n",
      "  Downloading importlib_resources-6.4.5-py3-none-any.whl.metadata (4.0 kB)\n",
      "Requirement already satisfied: numpy>=1.13 in /usr/local/art/.venv/lib/python3.12/site-packages (from cobra) (1.26.4)\n",
      "Collecting optlang~=1.8 (from cobra)\n",
      "  Downloading optlang-1.8.2-py2.py3-none-any.whl.metadata (8.1 kB)\n",
      "Requirement already satisfied: pandas<3.0,>=1.0 in /usr/local/art/.venv/lib/python3.12/site-packages (from cobra) (2.2.3)\n",
      "Requirement already satisfied: pydantic>=1.6 in /usr/local/art/.venv/lib/python3.12/site-packages (from cobra) (2.9.2)\n",
      "Collecting python-libsbml~=5.19 (from cobra)\n",
      "  Downloading python_libsbml-5.20.4-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl.metadata (532 bytes)\n",
      "Requirement already satisfied: rich>=8.0 in /usr/local/art/.venv/lib/python3.12/site-packages (from cobra) (13.8.1)\n",
      "Collecting ruamel.yaml~=0.16 (from cobra)\n",
      "  Downloading ruamel.yaml-0.18.6-py3-none-any.whl.metadata (23 kB)\n",
      "Collecting swiglpk (from cobra)\n",
      "  Downloading swiglpk-5.0.10-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl.metadata (5.5 kB)\n",
      "Requirement already satisfied: anyio in /usr/local/art/.venv/lib/python3.12/site-packages (from httpx~=0.24->cobra) (4.5.0)\n",
      "Requirement already satisfied: certifi in /usr/local/art/.venv/lib/python3.12/site-packages (from httpx~=0.24->cobra) (2024.8.30)\n",
      "Requirement already satisfied: httpcore==1.* in /usr/local/art/.venv/lib/python3.12/site-packages (from httpx~=0.24->cobra) (1.0.5)\n",
      "Requirement already satisfied: idna in /usr/local/art/.venv/lib/python3.12/site-packages (from httpx~=0.24->cobra) (3.10)\n",
      "Requirement already satisfied: sniffio in /usr/local/art/.venv/lib/python3.12/site-packages (from httpx~=0.24->cobra) (1.3.1)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in /usr/local/art/.venv/lib/python3.12/site-packages (from httpcore==1.*->httpx~=0.24->cobra) (0.14.0)\n",
      "Collecting sympy>=1.12.0 (from optlang~=1.8->cobra)\n",
      "  Downloading sympy-1.13.3-py3-none-any.whl.metadata (12 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/art/.venv/lib/python3.12/site-packages (from pandas<3.0,>=1.0->cobra) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /usr/local/art/.venv/lib/python3.12/site-packages (from pandas<3.0,>=1.0->cobra) (2024.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /usr/local/art/.venv/lib/python3.12/site-packages (from pandas<3.0,>=1.0->cobra) (2024.1)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/art/.venv/lib/python3.12/site-packages (from pydantic>=1.6->cobra) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.23.4 in /usr/local/art/.venv/lib/python3.12/site-packages (from pydantic>=1.6->cobra) (2.23.4)\n",
      "Requirement already satisfied: typing-extensions>=4.6.1 in /usr/local/art/.venv/lib/python3.12/site-packages (from pydantic>=1.6->cobra) (4.12.2)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/art/.venv/lib/python3.12/site-packages (from rich>=8.0->cobra) (3.0.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/art/.venv/lib/python3.12/site-packages (from rich>=8.0->cobra) (2.18.0)\n",
      "Collecting ruamel.yaml.clib>=0.2.7 (from ruamel.yaml~=0.16->cobra)\n",
      "  Downloading ruamel.yaml.clib-0.2.8-cp312-cp312-manylinux_2_24_aarch64.whl.metadata (2.2 kB)\n",
      "Requirement already satisfied: mdurl~=0.1 in /usr/local/art/.venv/lib/python3.12/site-packages (from markdown-it-py>=2.2.0->rich>=8.0->cobra) (0.1.2)\n",
      "Requirement already satisfied: six>=1.5 in /usr/local/art/.venv/lib/python3.12/site-packages (from python-dateutil>=2.8.2->pandas<3.0,>=1.0->cobra) (1.16.0)\n",
      "Collecting mpmath<1.4,>=1.1.0 (from sympy>=1.12.0->optlang~=1.8->cobra)\n",
      "  Downloading mpmath-1.3.0-py3-none-any.whl.metadata (8.6 kB)\n",
      "Downloading cobra-0.29.1-py2.py3-none-any.whl (1.2 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m12.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)\n",
      "Downloading depinfo-2.2.0-py3-none-any.whl (24 kB)\n",
      "Downloading diskcache-5.6.3-py3-none-any.whl (45 kB)\n",
      "Downloading optlang-1.8.2-py2.py3-none-any.whl (141 kB)\n",
      "Downloading python_libsbml-5.20.4-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (7.4 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7.4/7.4 MB\u001b[0m \u001b[31m3.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0mm\n",
      "\u001b[?25hDownloading ruamel.yaml-0.18.6-py3-none-any.whl (117 kB)\n",
      "Downloading swiglpk-5.0.10-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (2.3 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.3/2.3 MB\u001b[0m \u001b[31m4.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0mm\n",
      "\u001b[?25hDownloading future-1.0.0-py3-none-any.whl (491 kB)\n",
      "Downloading importlib_resources-6.4.5-py3-none-any.whl (36 kB)\n",
      "Downloading ruamel.yaml.clib-0.2.8-cp312-cp312-manylinux_2_24_aarch64.whl (643 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m643.9/643.9 kB\u001b[0m \u001b[31m557.8 kB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:--:--\u001b[0m\n",
      "\u001b[?25hDownloading sympy-1.13.3-py3-none-any.whl (6.2 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.2/6.2 MB\u001b[0m \u001b[31m3.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0mm\n",
      "\u001b[?25hDownloading mpmath-1.3.0-py3-none-any.whl (536 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m536.2/536.2 kB\u001b[0m \u001b[31m11.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: swiglpk, python-libsbml, mpmath, appdirs, sympy, ruamel.yaml.clib, importlib-resources, future, diskcache, depinfo, ruamel.yaml, optlang, cobra\n",
      "Successfully installed appdirs-1.4.4 cobra-0.29.1 depinfo-2.2.0 diskcache-5.6.3 future-1.0.0 importlib-resources-6.4.5 mpmath-1.3.0 optlang-1.8.2 python-libsbml-5.20.4 ruamel.yaml-0.18.6 ruamel.yaml.clib-0.2.8 swiglpk-5.0.10 sympy-1.13.3\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install cobra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3151ee35-e62b-4403-ab72-233da0388598",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import cobra\n",
    "model = cobra.io.load_json_model('../ml/cobrapy/data/iJN1463_IPP_bypass.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0f21d88d-b5ae-48a8-947a-f72f5d8d945e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP_0001\n"
     ]
    }
   ],
   "source": [
    "#Update one gene ID\n",
    "cobra.manipulation.modify.rename_genes(model = model, rename_dict = {'PP_s0001': 'PP_0001'})\n",
    "# model.genes.get_by_id('PP_s0001').id = 'PP_0001'\n",
    "print(model.genes.get_by_id('PP_0001'))\n",
    "gene_ids = [g.id for g in model.genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c9364ba2-0fe9-4417-8aea-e661773ab469",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_gsm_reactions(x, gene_ids):\n",
    "    if x in gene_ids:\n",
    "        return [r.id for r in model.genes.get_by_id(x).reactions]\n",
    "    else:\n",
    "        return None\n",
    "def get_gsm_rxn_subsystems(x, model):\n",
    "    subsystem_list = []\n",
    "    if x is not None:\n",
    "        for rxn_id in x:\n",
    "            rxn = model.reactions.get_by_id(rxn_id)\n",
    "            if rxn.subsystem is not None:\n",
    "                subsystem_list.append(rxn.subsystem)\n",
    "    return list(set(subsystem_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e3b52f4c-13ed-43c8-a9cf-71467486acbf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ATPS4rpp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>ATP synthase (four protons for one ATP) (periplasm)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0xffff5f1ef890</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>adp_c + 4.0 h_p + pi_c <=> atp_c + h2o_c + 3.0 h_c</p>\n",
       "                    <p style='text-align:right'>ADP C10H12N5O10P2 + 4.0 H+ + Phosphate <=> ATP C10H12N5O13P3 + H2O H2O + 3.0 H+</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>(PP_5417 and PP_5418 and PP_5419) and (PP_5412 and PP_5413 and PP_5414 and PP_5415 and PP_5416)...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ATPS4rpp at 0xffff5f1ef890>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('ATPS4rpp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6f1f4c35-12de-4013-af68-d11d2237e68d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "111"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge['in_gsm'] = df_merge['pp'].apply(lambda x: x in gene_ids)\n",
    "df_merge['gsm_rxns'] = df_merge['pp'].apply(lambda x: get_gsm_reactions(x, gene_ids))\n",
    "df_merge['gsm_subsystems'] = df_merge['gsm_rxns'].apply(lambda x: get_gsm_rxn_subsystems(x, model))\n",
    "df_merge['num_rxns'] = df_merge['gsm_rxns'].apply(lambda x: len(x) if x is not None else 0)\n",
    "df_merge['num_subsystems'] = df_merge['gsm_subsystems'].apply(lambda x: len(x) if x is not None else 0)\n",
    "df_merge['subsystem1'] = df_merge['gsm_subsystems'].apply(lambda x: x[0] if len(x)>0 else '-')\n",
    "df_merge['in_gsm'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d33d436f-5019-4be5-83ed-9037053a3faa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gsm_rxn</th>\n",
       "      <th>score</th>\n",
       "      <th>trend</th>\n",
       "      <th>ors?</th>\n",
       "      <th>subsystem</th>\n",
       "      <th>fluxDiff</th>\n",
       "      <th>FinalFlux</th>\n",
       "      <th>ands?</th>\n",
       "      <th>In</th>\n",
       "      <th>growth %</th>\n",
       "      <th>bgene</th>\n",
       "      <th>geneName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NADH16pp</td>\n",
       "      <td>8.598198e+06</td>\n",
       "      <td>Down</td>\n",
       "      <td>False</td>\n",
       "      <td>S_Oxidative_Phosphorylation</td>\n",
       "      <td>1.851036</td>\n",
       "      <td>4.839080e-01</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.895929</td>\n",
       "      <td>PP_4119 and PP_4120 and PP_4121 and PP_4122 an...</td>\n",
       "      <td>['nuoA', 'nuoB', 'nuoC', 'nuoE', 'nuoF', 'nuoG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CS</td>\n",
       "      <td>8.325106e+03</td>\n",
       "      <td>Down</td>\n",
       "      <td>False</td>\n",
       "      <td>S_TCA_Cycle</td>\n",
       "      <td>0.611303</td>\n",
       "      <td>-9.530220e-14</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.963479</td>\n",
       "      <td>PP_4194</td>\n",
       "      <td>['gltA']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CBMKr</td>\n",
       "      <td>5.883053e+03</td>\n",
       "      <td>Down</td>\n",
       "      <td>False</td>\n",
       "      <td>S_Glutamate_Metabolism</td>\n",
       "      <td>0.064952</td>\n",
       "      <td>-2.444027e-13</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.993853</td>\n",
       "      <td>PP_0999</td>\n",
       "      <td>['arcC']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IMPD</td>\n",
       "      <td>7.857951e+02</td>\n",
       "      <td>Down</td>\n",
       "      <td>False</td>\n",
       "      <td>S_Purine_Metabolism</td>\n",
       "      <td>0.017844</td>\n",
       "      <td>4.895565e-13</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.997936</td>\n",
       "      <td>PP_1031</td>\n",
       "      <td>['guaB']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PPND</td>\n",
       "      <td>7.697478e+02</td>\n",
       "      <td>Down</td>\n",
       "      <td>False</td>\n",
       "      <td>S_Phenylalanine_Tyrosine_Tryptophan_Biosynthesis</td>\n",
       "      <td>0.011026</td>\n",
       "      <td>-4.048215e-14</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>0.997860</td>\n",
       "      <td>PP_1770</td>\n",
       "      <td>['']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gsm_rxn         score trend   ors?  \\\n",
       "0  NADH16pp  8.598198e+06  Down  False   \n",
       "1        CS  8.325106e+03  Down  False   \n",
       "2     CBMKr  5.883053e+03  Down  False   \n",
       "3      IMPD  7.857951e+02  Down  False   \n",
       "4      PPND  7.697478e+02  Down  False   \n",
       "\n",
       "                                          subsystem  fluxDiff     FinalFlux  \\\n",
       "0                       S_Oxidative_Phosphorylation  1.851036  4.839080e-01   \n",
       "1                                       S_TCA_Cycle  0.611303 -9.530220e-14   \n",
       "2                            S_Glutamate_Metabolism  0.064952 -2.444027e-13   \n",
       "3                               S_Purine_Metabolism  0.017844  4.895565e-13   \n",
       "4  S_Phenylalanine_Tyrosine_Tryptophan_Biosynthesis  0.011026 -4.048215e-14   \n",
       "\n",
       "   ands?    In  growth %                                              bgene  \\\n",
       "0   True  True  0.895929  PP_4119 and PP_4120 and PP_4121 and PP_4122 an...   \n",
       "1  False  True  0.963479                                            PP_4194   \n",
       "2  False  True  0.993853                                            PP_0999   \n",
       "3  False  True  0.997936                                            PP_1031   \n",
       "4  False  True  0.997860                                            PP_1770   \n",
       "\n",
       "                                            geneName  \n",
       "0  ['nuoA', 'nuoB', 'nuoC', 'nuoE', 'nuoF', 'nuoG...  \n",
       "1                                           ['gltA']  \n",
       "2                                           ['arcC']  \n",
       "3                                           ['guaB']  \n",
       "4                                               ['']  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_retap = pd.read_csv('../ml/cobrapy/data/CRISPRiRecommendations_6_15_22.csv').rename(columns = {'Unnamed: 0': 'gsm_rxn'})\n",
    "df_retap.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a7626d5d-f896-4b39-9650-abbb8bc0ce9d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PP_0001', 'PP_0103', 'PP_0104', 'PP_0105', 'PP_0106', 'PP_0225', 'PP_0226', 'PP_0227', 'PP_0356', 'PP_0658', 'PP_0691', 'PP_0812', 'PP_0813', 'PP_0814', 'PP_0815', 'PP_0999', 'PP_1023', 'PP_1031', 'PP_1240', 'PP_1317', 'PP_1318', 'PP_1319', 'PP_1457', 'PP_1506', 'PP_1610', 'PP_1620', 'PP_1664', 'PP_1769', 'PP_1770', 'PP_1771', 'PP_1914', 'PP_2095', 'PP_4043', 'PP_4116', 'PP_4119', 'PP_4120', 'PP_4121', 'PP_4122', 'PP_4123', 'PP_4124', 'PP_4125', 'PP_4126', 'PP_4127', 'PP_4128', 'PP_4129', 'PP_4130', 'PP_4131', 'PP_4159', 'PP_4160', 'PP_4161', 'PP_4174', 'PP_4175', 'PP_4187', 'PP_4188', 'PP_4189', 'PP_4194', 'PP_4266', 'PP_4635', 'PP_4650', 'PP_4651', 'PP_4862', 'PP_5046', 'PP_5128', 'PP_5186', 'PP_5203']\n",
      "RETAP genes: N = 65\n"
     ]
    }
   ],
   "source": [
    "def get_gene_list(x):\n",
    "    pp_list = []\n",
    "    if x is not np.nan:\n",
    "        pp_list = re.findall('PP_[0-9]{4}', x)\n",
    "    return pp_list\n",
    "df_retap['gene_list'] = df_retap['bgene'].apply(get_gene_list)\n",
    "retap_genes = list(set([g for gene_list in df_retap['gene_list'].values for g in gene_list]))\n",
    "#This target is misnamed in the GSM (named PP_s0001) so our regex doesn't pick it up\n",
    "retap_genes.append('PP_0001')\n",
    "retap_genes.sort()\n",
    "print(retap_genes)\n",
    "print(f'RETAP genes: N = {len(retap_genes)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "09020ee6-0751-4699-a3c5-47991f8fc7cc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RETAP genes in target list: N = 54\n"
     ]
    }
   ],
   "source": [
    "df_merge['retap_target'] = df_merge['pp'].apply(lambda x: x in retap_genes)\n",
    "print(f\"RETAP genes in target list: N = {sum(df_merge['retap_target'])}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3459809c-312b-41d0-88c4-02e75171e08b",
   "metadata": {},
   "source": [
    "## Get heuristic information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0d61500b-36f7-444c-b078-d65d98ecc3fb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pp</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP_0338</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP_0362</td>\n",
       "      <td>Metabolism of cofactors and vitamins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP_0437</td>\n",
       "      <td>Metabolism of cofactors and vitamins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PP_0548</td>\n",
       "      <td>Metabolism of cofactors and vitamins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PP_0582</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PP_0774</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PP_1157</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PP_1444</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PP_1607</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PP_1777</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PP_2137</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PP_2137</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PP_2213</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PP_2589</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PP_3282</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PP_3355</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PP_3394</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PP_3578</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PP_3754</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PP_3755</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PP_4194</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PP_4487</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PP_4549</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PP_4550</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PP_4636</td>\n",
       "      <td>Lipid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PP_4678</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PP_4679</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PP_4702</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PP_5064</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PP_5288</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PP_5335</td>\n",
       "      <td>Nucleotide metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PP_5416</td>\n",
       "      <td>ATP Synthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PP_5417</td>\n",
       "      <td>ATP Synthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>PP_5419</td>\n",
       "      <td>ATP Synthesis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         pp                              category\n",
       "0   PP_0338               Carbohydrate metabolism\n",
       "1   PP_0362  Metabolism of cofactors and vitamins\n",
       "2   PP_0437  Metabolism of cofactors and vitamins\n",
       "3   PP_0548  Metabolism of cofactors and vitamins\n",
       "4   PP_0582                      Lipid metabolism\n",
       "5   PP_0774               Carbohydrate metabolism\n",
       "6   PP_1157                 Amino acid metabolism\n",
       "7   PP_1444               Carbohydrate metabolism\n",
       "8   PP_1607                      Lipid metabolism\n",
       "9   PP_1777               Carbohydrate metabolism\n",
       "10  PP_2137                 Amino acid metabolism\n",
       "11  PP_2137                      Lipid metabolism\n",
       "12  PP_2213               Carbohydrate metabolism\n",
       "13  PP_2589                 Amino acid metabolism\n",
       "14  PP_3282                      Lipid metabolism\n",
       "15  PP_3355                      Lipid metabolism\n",
       "16  PP_3394                      Lipid metabolism\n",
       "17  PP_3578               Carbohydrate metabolism\n",
       "18  PP_3754                      Lipid metabolism\n",
       "19  PP_3755                      Lipid metabolism\n",
       "20  PP_4194               Carbohydrate metabolism\n",
       "21  PP_4487               Carbohydrate metabolism\n",
       "22  PP_4549                      Lipid metabolism\n",
       "23  PP_4550                      Lipid metabolism\n",
       "24  PP_4636                      Lipid metabolism\n",
       "25  PP_4678                 Amino acid metabolism\n",
       "26  PP_4679                 Amino acid metabolism\n",
       "27  PP_4702               Carbohydrate metabolism\n",
       "28  PP_5064                 Amino acid metabolism\n",
       "29  PP_5288               Carbohydrate metabolism\n",
       "30  PP_5335                 Nucleotide metabolism\n",
       "31  PP_5416                         ATP Synthesis\n",
       "32  PP_5417                         ATP Synthesis\n",
       "33  PP_5419                         ATP Synthesis"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_heuristic = pd.read_csv('../ml/data/heuristic_target_categories.csv')\n",
    "df_heuristic"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ddb66f0-0765-4cef-b263-194499954c3c",
   "metadata": {},
   "source": [
    "## Get target expression/isoprenol data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ae2f995c-6746-4a95-840d-8e8663802c06",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pp</th>\n",
       "      <th>Entry_x</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Gene Ontology (molecular function)</th>\n",
       "      <th>Gene Ontology (GO)</th>\n",
       "      <th>Gene Ontology IDs</th>\n",
       "      <th>Gene Ontology (cellular component)</th>\n",
       "      <th>...</th>\n",
       "      <th>gsm_rxns</th>\n",
       "      <th>gsm_subsystems</th>\n",
       "      <th>num_rxns</th>\n",
       "      <th>num_subsystems</th>\n",
       "      <th>subsystem1</th>\n",
       "      <th>retap_target</th>\n",
       "      <th>isoprenol_mean</th>\n",
       "      <th>dCas9_expr</th>\n",
       "      <th>pass_dbtl0_filter</th>\n",
       "      <th>target_control_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP_0001</td>\n",
       "      <td>P0A151</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>PARB_PSEPK</td>\n",
       "      <td>Probable chromosome-partitioning protein ParB</td>\n",
       "      <td>parB PP_0001</td>\n",
       "      <td>DNA binding [GO:0003677]</td>\n",
       "      <td>DNA binding [GO:0003677]; chromosome segregati...</td>\n",
       "      <td>GO:0003677; GO:0007059</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[TNTHAC2, Cotpp, ETOHtrpp, MEOHtex, FE3PYOVDL2...</td>\n",
       "      <td>[Unassigned, S_Alternate_Carbon_and_Nitrogen_s...</td>\n",
       "      <td>41</td>\n",
       "      <td>15</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>True</td>\n",
       "      <td>58.026153</td>\n",
       "      <td>1.037982</td>\n",
       "      <td>False</td>\n",
       "      <td>0.077754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP_0103</td>\n",
       "      <td>Q88RM6</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Q88RM6_PSEPK</td>\n",
       "      <td>Cytochrome c oxidase subunit 2 (EC 7.1.1.9)</td>\n",
       "      <td>PP_0103</td>\n",
       "      <td>copper ion binding [GO:0005507]; cytochrome-c ...</td>\n",
       "      <td>plasma membrane [GO:0005886]; respirasome [GO:...</td>\n",
       "      <td>GO:0004129; GO:0005507; GO:0005886; GO:0020037...</td>\n",
       "      <td>plasma membrane [GO:0005886]; respirasome [GO:...</td>\n",
       "      <td>...</td>\n",
       "      <td>[CYTCAA3pp]</td>\n",
       "      <td>[S_Oxidative_Phosphorylation]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S_Oxidative_Phosphorylation</td>\n",
       "      <td>True</td>\n",
       "      <td>183.521667</td>\n",
       "      <td>0.199516</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP_0104</td>\n",
       "      <td>Q88RM5</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Q88RM5_PSEPK</td>\n",
       "      <td>Cytochrome c oxidase subunit 1 (EC 7.1.1.9)</td>\n",
       "      <td>ctaD PP_0104</td>\n",
       "      <td>cytochrome-c oxidase activity [GO:0004129]; he...</td>\n",
       "      <td>plasma membrane [GO:0005886]; respiratory chai...</td>\n",
       "      <td>GO:0004129; GO:0005886; GO:0006119; GO:0015990...</td>\n",
       "      <td>plasma membrane [GO:0005886]; respiratory chai...</td>\n",
       "      <td>...</td>\n",
       "      <td>[CYTCAA3pp]</td>\n",
       "      <td>[S_Oxidative_Phosphorylation]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S_Oxidative_Phosphorylation</td>\n",
       "      <td>True</td>\n",
       "      <td>130.414757</td>\n",
       "      <td>1.111125</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        pp Entry_x    Reviewed    Entry Name  \\\n",
       "0  PP_0001  P0A151    reviewed    PARB_PSEPK   \n",
       "1  PP_0103  Q88RM6  unreviewed  Q88RM6_PSEPK   \n",
       "2  PP_0104  Q88RM5  unreviewed  Q88RM5_PSEPK   \n",
       "\n",
       "                                   Protein names    Gene Names  \\\n",
       "0  Probable chromosome-partitioning protein ParB  parB PP_0001   \n",
       "1    Cytochrome c oxidase subunit 2 (EC 7.1.1.9)       PP_0103   \n",
       "2    Cytochrome c oxidase subunit 1 (EC 7.1.1.9)  ctaD PP_0104   \n",
       "\n",
       "                  Gene Ontology (molecular function)  \\\n",
       "0                           DNA binding [GO:0003677]   \n",
       "1  copper ion binding [GO:0005507]; cytochrome-c ...   \n",
       "2  cytochrome-c oxidase activity [GO:0004129]; he...   \n",
       "\n",
       "                                  Gene Ontology (GO)  \\\n",
       "0  DNA binding [GO:0003677]; chromosome segregati...   \n",
       "1  plasma membrane [GO:0005886]; respirasome [GO:...   \n",
       "2  plasma membrane [GO:0005886]; respiratory chai...   \n",
       "\n",
       "                                   Gene Ontology IDs  \\\n",
       "0                             GO:0003677; GO:0007059   \n",
       "1  GO:0004129; GO:0005507; GO:0005886; GO:0020037...   \n",
       "2  GO:0004129; GO:0005886; GO:0006119; GO:0015990...   \n",
       "\n",
       "                  Gene Ontology (cellular component)  ...  \\\n",
       "0                                                NaN  ...   \n",
       "1  plasma membrane [GO:0005886]; respirasome [GO:...  ...   \n",
       "2  plasma membrane [GO:0005886]; respiratory chai...  ...   \n",
       "\n",
       "                                            gsm_rxns  \\\n",
       "0  [TNTHAC2, Cotpp, ETOHtrpp, MEOHtex, FE3PYOVDL2...   \n",
       "1                                        [CYTCAA3pp]   \n",
       "2                                        [CYTCAA3pp]   \n",
       "\n",
       "                                      gsm_subsystems num_rxns num_subsystems  \\\n",
       "0  [Unassigned, S_Alternate_Carbon_and_Nitrogen_s...       41             15   \n",
       "1                      [S_Oxidative_Phosphorylation]        1              1   \n",
       "2                      [S_Oxidative_Phosphorylation]        1              1   \n",
       "\n",
       "                    subsystem1 retap_target isoprenol_mean dCas9_expr  \\\n",
       "0                   Unassigned         True      58.026153   1.037982   \n",
       "1  S_Oxidative_Phosphorylation         True     183.521667   0.199516   \n",
       "2  S_Oxidative_Phosphorylation         True     130.414757   1.111125   \n",
       "\n",
       "  pass_dbtl0_filter target_control_ratio  \n",
       "0             False             0.077754  \n",
       "1             False                  NaN  \n",
       "2             False                  NaN  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge['isoprenol_mean'] = df_merge['pp'].apply(lambda x: df_target.loc[df_target['Target'] == x, 'Isoprenol mean'].values[0])\n",
    "df_merge['dCas9_expr'] = df_merge['pp'].apply(lambda x: df_target.loc[df_target['Target'] == x, 'dCas9'].values[0])\n",
    "df_merge['pass_dbtl0_filter'] = df_merge['pp'].apply(lambda x: df_target.loc[df_target['Target'] == x, 'to_keep'].values[0])\n",
    "df_merge['target_control_ratio'] = df_merge['pp'].apply(lambda x: df_target.loc[df_target['Target'] == x, 'Target:Control'].values[0])\n",
    "df_merge.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e026c5c2-c683-4cc6-8de4-16516f5084a8",
   "metadata": {},
   "source": [
    "## Get DBTL3 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d596c19c-fad2-4493-98fb-7e36e3f808c9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N = 16 dbtl3 genes\n",
      "['PP_1317', 'PP_0751', 'PP_4120', 'PP_0813', 'PP_0437', 'PP_1506', 'PP_4191', 'PP_1769', 'PP_2136', 'PP_0528', 'PP_0814', 'PP_0815', 'PP_4189', 'PP_0812', 'PP_4192', 'PP_0368']\n"
     ]
    }
   ],
   "source": [
    "dbtl3_data = pd.read_pickle('../dbtl3ONLY_figures_and_analysis/dbtl3ONLY_isoprenol_processed_df.pkl')\n",
    "# dbtl3_all_genes = []\n",
    "dbtl3_all_genes = [gene_list for gene_list in dbtl3_data['sorted_gene_list']]\n",
    "dbtl3_all_genes = list(set([gene for gene_list in dbtl3_all_genes for gene in gene_list]))\n",
    "print(f'N = {len(dbtl3_all_genes)} dbtl3 genes')\n",
    "print(dbtl3_all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "21488781-5226-43d3-aa3d-3577788996a4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_merge['pass_dbtl3'] = df_merge['pp'].apply(lambda x: x in dbtl3_all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e83d9e39-6662-4d2b-9efe-63e72d714a29",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['pp', 'Entry_x', 'Reviewed', 'Entry Name', 'Protein names',\n",
       "       'Gene Names', 'Gene Ontology (molecular function)',\n",
       "       'Gene Ontology (GO)', 'Gene Ontology IDs',\n",
       "       'Gene Ontology (cellular component)',\n",
       "       'Gene Ontology (biological process)', 'New_tag', 'Preferred_name',\n",
       "       'Nickname', 'Description', 'COG', 'GOs', 'EC', 'KEGG_ko',\n",
       "       'KEGG_Reaction', 'pref_first3', 'in_gsm', 'gsm_rxns', 'gsm_subsystems',\n",
       "       'num_rxns', 'num_subsystems', 'subsystem1', 'retap_target',\n",
       "       'isoprenol_mean', 'dCas9_expr', 'pass_dbtl0_filter',\n",
       "       'target_control_ratio', 'pass_dbtl3'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "339a7a7b-6b3c-4485-8a01-35b90c518117",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_merge.to_csv('target_info_output/merged_uniprot_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "37707882-14c3-4397-8f28-9c4e3ae4c852",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pp</th>\n",
       "      <th>Entry_x</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Gene Ontology (molecular function)</th>\n",
       "      <th>Gene Ontology (GO)</th>\n",
       "      <th>Gene Ontology IDs</th>\n",
       "      <th>Gene Ontology (cellular component)</th>\n",
       "      <th>...</th>\n",
       "      <th>gsm_subsystems</th>\n",
       "      <th>num_rxns</th>\n",
       "      <th>num_subsystems</th>\n",
       "      <th>subsystem1</th>\n",
       "      <th>retap_target</th>\n",
       "      <th>isoprenol_mean</th>\n",
       "      <th>dCas9_expr</th>\n",
       "      <th>pass_dbtl0_filter</th>\n",
       "      <th>target_control_ratio</th>\n",
       "      <th>pass_dbtl3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP_0001</td>\n",
       "      <td>P0A151</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>PARB_PSEPK</td>\n",
       "      <td>Probable chromosome-partitioning protein ParB</td>\n",
       "      <td>parB PP_0001</td>\n",
       "      <td>DNA binding [GO:0003677]</td>\n",
       "      <td>DNA binding [GO:0003677]; chromosome segregati...</td>\n",
       "      <td>GO:0003677; GO:0007059</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[Unassigned, S_Alternate_Carbon_and_Nitrogen_s...</td>\n",
       "      <td>41</td>\n",
       "      <td>15</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>True</td>\n",
       "      <td>58.026153</td>\n",
       "      <td>1.037982</td>\n",
       "      <td>False</td>\n",
       "      <td>0.077754</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP_0103</td>\n",
       "      <td>Q88RM6</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Q88RM6_PSEPK</td>\n",
       "      <td>Cytochrome c oxidase subunit 2 (EC 7.1.1.9)</td>\n",
       "      <td>PP_0103</td>\n",
       "      <td>copper ion binding [GO:0005507]; cytochrome-c ...</td>\n",
       "      <td>plasma membrane [GO:0005886]; respirasome [GO:...</td>\n",
       "      <td>GO:0004129; GO:0005507; GO:0005886; GO:0020037...</td>\n",
       "      <td>plasma membrane [GO:0005886]; respirasome [GO:...</td>\n",
       "      <td>...</td>\n",
       "      <td>[S_Oxidative_Phosphorylation]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S_Oxidative_Phosphorylation</td>\n",
       "      <td>True</td>\n",
       "      <td>183.521667</td>\n",
       "      <td>0.199516</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP_0104</td>\n",
       "      <td>Q88RM5</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Q88RM5_PSEPK</td>\n",
       "      <td>Cytochrome c oxidase subunit 1 (EC 7.1.1.9)</td>\n",
       "      <td>ctaD PP_0104</td>\n",
       "      <td>cytochrome-c oxidase activity [GO:0004129]; he...</td>\n",
       "      <td>plasma membrane [GO:0005886]; respiratory chai...</td>\n",
       "      <td>GO:0004129; GO:0005886; GO:0006119; GO:0015990...</td>\n",
       "      <td>plasma membrane [GO:0005886]; respiratory chai...</td>\n",
       "      <td>...</td>\n",
       "      <td>[S_Oxidative_Phosphorylation]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S_Oxidative_Phosphorylation</td>\n",
       "      <td>True</td>\n",
       "      <td>130.414757</td>\n",
       "      <td>1.111125</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        pp Entry_x    Reviewed    Entry Name  \\\n",
       "0  PP_0001  P0A151    reviewed    PARB_PSEPK   \n",
       "1  PP_0103  Q88RM6  unreviewed  Q88RM6_PSEPK   \n",
       "2  PP_0104  Q88RM5  unreviewed  Q88RM5_PSEPK   \n",
       "\n",
       "                                   Protein names    Gene Names  \\\n",
       "0  Probable chromosome-partitioning protein ParB  parB PP_0001   \n",
       "1    Cytochrome c oxidase subunit 2 (EC 7.1.1.9)       PP_0103   \n",
       "2    Cytochrome c oxidase subunit 1 (EC 7.1.1.9)  ctaD PP_0104   \n",
       "\n",
       "                  Gene Ontology (molecular function)  \\\n",
       "0                           DNA binding [GO:0003677]   \n",
       "1  copper ion binding [GO:0005507]; cytochrome-c ...   \n",
       "2  cytochrome-c oxidase activity [GO:0004129]; he...   \n",
       "\n",
       "                                  Gene Ontology (GO)  \\\n",
       "0  DNA binding [GO:0003677]; chromosome segregati...   \n",
       "1  plasma membrane [GO:0005886]; respirasome [GO:...   \n",
       "2  plasma membrane [GO:0005886]; respiratory chai...   \n",
       "\n",
       "                                   Gene Ontology IDs  \\\n",
       "0                             GO:0003677; GO:0007059   \n",
       "1  GO:0004129; GO:0005507; GO:0005886; GO:0020037...   \n",
       "2  GO:0004129; GO:0005886; GO:0006119; GO:0015990...   \n",
       "\n",
       "                  Gene Ontology (cellular component)  ...  \\\n",
       "0                                                NaN  ...   \n",
       "1  plasma membrane [GO:0005886]; respirasome [GO:...  ...   \n",
       "2  plasma membrane [GO:0005886]; respiratory chai...  ...   \n",
       "\n",
       "                                      gsm_subsystems num_rxns num_subsystems  \\\n",
       "0  [Unassigned, S_Alternate_Carbon_and_Nitrogen_s...       41             15   \n",
       "1                      [S_Oxidative_Phosphorylation]        1              1   \n",
       "2                      [S_Oxidative_Phosphorylation]        1              1   \n",
       "\n",
       "                    subsystem1 retap_target isoprenol_mean dCas9_expr  \\\n",
       "0                   Unassigned         True      58.026153   1.037982   \n",
       "1  S_Oxidative_Phosphorylation         True     183.521667   0.199516   \n",
       "2  S_Oxidative_Phosphorylation         True     130.414757   1.111125   \n",
       "\n",
       "  pass_dbtl0_filter target_control_ratio pass_dbtl3  \n",
       "0             False             0.077754      False  \n",
       "1             False                  NaN      False  \n",
       "2             False                  NaN      False  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec4b13a3-554e-4dcd-bcc0-e5c51fbc7962",
   "metadata": {},
   "source": [
    "## Get Rb-TnSeq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "468c05d3-804d-4efc-938d-14cff125dc68",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_rbtnseq_data(x):\n",
    "    vals = rbtnseq_df.loc[rbtnseq_df['locusId'] == x, rbtnseq_condition].values\n",
    "    if len(vals) == 1:\n",
    "        return vals[0]\n",
    "    elif len(vals) == 0:\n",
    "        return 'No value'\n",
    "    else: \n",
    "        return 'Multiple Values'\n",
    "rbtnseq_df = pd.read_csv('rbtnseq_data_m9only.tsv', sep='\\t')\n",
    "rbtnseq_condition = 'set8IT059 Adaptation in M9 media timecourse 1; day 3'\n",
    "df_merge['rbtnseq_val'] = df_merge['pp'].apply(get_rbtnseq_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "607ab06a-48df-4aa9-b31c-ca9667a55532",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pp</th>\n",
       "      <th>Entry_x</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Gene Ontology (molecular function)</th>\n",
       "      <th>Gene Ontology (GO)</th>\n",
       "      <th>Gene Ontology IDs</th>\n",
       "      <th>Gene Ontology (cellular component)</th>\n",
       "      <th>...</th>\n",
       "      <th>num_rxns</th>\n",
       "      <th>num_subsystems</th>\n",
       "      <th>subsystem1</th>\n",
       "      <th>retap_target</th>\n",
       "      <th>isoprenol_mean</th>\n",
       "      <th>dCas9_expr</th>\n",
       "      <th>pass_dbtl0_filter</th>\n",
       "      <th>target_control_ratio</th>\n",
       "      <th>pass_dbtl3</th>\n",
       "      <th>rbtnseq_val</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PP_0814</td>\n",
       "      <td>Q88PN5</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Q88PN5_PSEPK</td>\n",
       "      <td>Cytochrome bo(3) ubiquinol oxidase subunit 3 (...</td>\n",
       "      <td>cyoC PP_0814</td>\n",
       "      <td>cytochrome bo3 ubiquinol oxidase activity [GO:...</td>\n",
       "      <td>plasma membrane [GO:0005886]; cytochrome bo3 u...</td>\n",
       "      <td>GO:0004129; GO:0005886; GO:0009486; GO:0019646</td>\n",
       "      <td>plasma membrane [GO:0005886]</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S_Oxidative_Phosphorylation</td>\n",
       "      <td>True</td>\n",
       "      <td>312.395913</td>\n",
       "      <td>0.026424</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>No value</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         pp Entry_x    Reviewed    Entry Name  \\\n",
       "25  PP_0814  Q88PN5  unreviewed  Q88PN5_PSEPK   \n",
       "\n",
       "                                        Protein names    Gene Names  \\\n",
       "25  Cytochrome bo(3) ubiquinol oxidase subunit 3 (...  cyoC PP_0814   \n",
       "\n",
       "                   Gene Ontology (molecular function)  \\\n",
       "25  cytochrome bo3 ubiquinol oxidase activity [GO:...   \n",
       "\n",
       "                                   Gene Ontology (GO)  \\\n",
       "25  plasma membrane [GO:0005886]; cytochrome bo3 u...   \n",
       "\n",
       "                                 Gene Ontology IDs  \\\n",
       "25  GO:0004129; GO:0005886; GO:0009486; GO:0019646   \n",
       "\n",
       "   Gene Ontology (cellular component)  ... num_rxns num_subsystems  \\\n",
       "25       plasma membrane [GO:0005886]  ...        1              1   \n",
       "\n",
       "                     subsystem1 retap_target isoprenol_mean dCas9_expr  \\\n",
       "25  S_Oxidative_Phosphorylation         True     312.395913   0.026424   \n",
       "\n",
       "   pass_dbtl0_filter target_control_ratio pass_dbtl3 rbtnseq_val  \n",
       "25              True                  NaN       True    No value  \n",
       "\n",
       "[1 rows x 34 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge[df_merge['pp'] == 'PP_0814']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5fe4560e-e824-4223-b075-db3a8085139d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set10IT006 L-Arginine (C)\n"
     ]
    }
   ],
   "source": [
    "rbtnseq_df_full = pd.read_csv('rbtnseq_data_all.tsv', sep = '\\t')\n",
    "cond_columns = rbtnseq_df_full.columns[5:]\n",
    "print(cond_columns[0])\n",
    "sum([x in rbtnseq_df_full['locusId'].values for x in df_merge['pp'].values])\n",
    "df_merge['in_rbtnseq'] = df_merge['pp'].apply(lambda x: x in rbtnseq_df_full['locusId'].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab7ff1df-ec24-4d0f-a588-4ecedcb9d8bc",
   "metadata": {},
   "source": [
    "## Cross reference DBTL0 filter with DBTL1 targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c011b9d1-c22f-4d5b-9b77-1f6c8ebd0428",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dbtl1_df = pd.read_pickle('../dbtl1_figures_and_analysis/dbtl0_dbtl1_isoprenol_processed_df.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "489fa6f4-8e34-4337-a4dd-ef4d4dccad36",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dbtl1_genes = list(\n",
    "    set(\n",
    "        [g for sublist in dbtl1_df.loc[dbtl1_df['cycle'] == 1, 'sorted_gene_list'].values for g in sublist]\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "79aeff3e-754b-4e75-ad87-f85915450125",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      False\n",
       "1      False\n",
       "2      False\n",
       "3      False\n",
       "4      False\n",
       "       ...  \n",
       "115    False\n",
       "116    False\n",
       "117     True\n",
       "118    False\n",
       "119    False\n",
       "Name: pass_dbtl0_filter, Length: 120, dtype: bool"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge['pass_dbtl0_filter']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3fd5d731-2c37-4e9d-a799-8bbeb4716a74",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>in_dbtl1</th>\n",
       "      <th>pass_dbtl0_filter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     in_dbtl1  pass_dbtl0_filter\n",
       "0       False              False\n",
       "1       False              False\n",
       "2       False              False\n",
       "3       False              False\n",
       "4       False              False\n",
       "..        ...                ...\n",
       "115     False              False\n",
       "116     False              False\n",
       "117     False               True\n",
       "118     False              False\n",
       "119     False              False\n",
       "\n",
       "[120 rows x 2 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge['in_dbtl1'] = df_merge['pp'].apply(lambda x: x in dbtl1_genes)\n",
    "\n",
    "df_merge[['in_dbtl1', 'pass_dbtl0_filter']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2103d54c-c0eb-4add-a06a-a1b8517c7c47",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pp</th>\n",
       "      <th>pass_dbtl0_filter</th>\n",
       "      <th>in_dbtl1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PP_0168</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PP_0654</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>PP_1318</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>PP_1755</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>PP_2925</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>PP_4123</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>PP_4185</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>PP_4487</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>PP_4636</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>PP_5203</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PP_5417</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          pp  pass_dbtl0_filter  in_dbtl1\n",
       "5    PP_0168               True     False\n",
       "17   PP_0654               True     False\n",
       "36   PP_1318               True     False\n",
       "46   PP_1755               True     False\n",
       "58   PP_2925               True     False\n",
       "78   PP_4123               True     False\n",
       "86   PP_4185               True     False\n",
       "96   PP_4487               True     False\n",
       "100  PP_4636               True     False\n",
       "113  PP_5203               True     False\n",
       "117  PP_5417               True     False"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge.loc[df_merge['in_dbtl1']!=df_merge['pass_dbtl0_filter'], ['pp', 'pass_dbtl0_filter', 'in_dbtl1']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8abe036b-f683-4458-a5ed-4a9a3c866f99",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Make target list\n",
    "Here, I make a csv file containing information about the initial targets we chose:\n",
    "- Locus name (PP_XXXX) - done\n",
    "- Protein name - From uniprot - done\n",
    "- Protein function - From uniprot - done\n",
    "- Source (heuristic vs. Flux-RETAP) - From Ian's docs - done\n",
    "- RbTnSeq essentiallity\n",
    "- Isoprenol titer - From DBTL0 results - done\n",
    "- dCas9 expression normalized to control - from DBTL0 results - done\n",
    "- Whether it passed filters - From DBTL0 results - done\n",
    "- Whether it was included after DBTL3 - From DBTL3 results\n",
    "- Anything else? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "09cc83eb-6c04-4a2c-b1e6-8399df7c1846",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "columns_to_keep = [\n",
    "    'pp',\n",
    "    'Protein names',\n",
    "    'retap_target',\n",
    "    'in_rbtnseq',\n",
    "    'isoprenol_mean', \n",
    "    'dCas9_expr',\n",
    "    'target_control_ratio',\n",
    "    'pass_dbtl0_filter',\n",
    "    'pass_dbtl3'\n",
    "]\n",
    "df_table = df_merge[columns_to_keep].copy()\n",
    "\n",
    "df_table['retap_target'] = df_table['retap_target'].apply(lambda x: 'FR' if x else 'H')\n",
    "df_table['in_rbtnseq'] = df_table['in_rbtnseq'].apply(lambda x: 'N' if x else 'Y') \n",
    "df_table['isoprenol_mean'] = np.round(df_table['isoprenol_mean'], 2)\n",
    "df_table['dCas9_expr'] = np.round(df_table['dCas9_expr'], 2)\n",
    "df_table['target_control_ratio'] = np.round(df_table['target_control_ratio'], 2)\n",
    "rename_dict = {\n",
    "    'pp': 'Locus number',\n",
    "    'Protein names': 'Protein name',\n",
    "    'retap_target': 'Source (Flux-RETAP or Heuristic)',\n",
    "    'isoprenol_mean': 'Mean isoprenol titer (mg/L)',\n",
    "    'dCas9_expr': 'dCas9/control',\n",
    "    'target_control_ratio': 'Target/control',\n",
    "    'pass_dbtl0_filter': 'Passed DBTL0 filter?',\n",
    "    'pass_dbtl3': 'Used after DBTL2/3?',\n",
    "    'in_rbtnseq': 'RbTn-Seq Essential?'\n",
    "}\n",
    "\n",
    "df_table = df_table.rename(columns = rename_dict)\n",
    "\n",
    "df_table.to_csv('Target_table.csv', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "9750ba35-903b-4513-8bb8-880fba38fe89",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Locus number</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>Source (Flux-RETAP or Heuristic)</th>\n",
       "      <th>RbTn-Seq Essential?</th>\n",
       "      <th>Mean isoprenol titer (mg/L)</th>\n",
       "      <th>dCas9/control</th>\n",
       "      <th>Target/control</th>\n",
       "      <th>Passed DBTL0 filter?</th>\n",
       "      <th>Used after DBTL2/3?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP_0001</td>\n",
       "      <td>Probable chromosome-partitioning protein ParB</td>\n",
       "      <td>FR</td>\n",
       "      <td>N</td>\n",
       "      <td>58.03</td>\n",
       "      <td>1.04</td>\n",
       "      <td>0.08</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP_0103</td>\n",
       "      <td>Cytochrome c oxidase subunit 2 (EC 7.1.1.9)</td>\n",
       "      <td>FR</td>\n",
       "      <td>N</td>\n",
       "      <td>183.52</td>\n",
       "      <td>0.20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP_0104</td>\n",
       "      <td>Cytochrome c oxidase subunit 1 (EC 7.1.1.9)</td>\n",
       "      <td>FR</td>\n",
       "      <td>N</td>\n",
       "      <td>130.41</td>\n",
       "      <td>1.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PP_0105</td>\n",
       "      <td>Cytochrome c oxidase assembly protein CtaG</td>\n",
       "      <td>FR</td>\n",
       "      <td>N</td>\n",
       "      <td>144.84</td>\n",
       "      <td>0.96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PP_0106</td>\n",
       "      <td>cytochrome-c oxidase (EC 7.1.1.9) (Cytochrome ...</td>\n",
       "      <td>FR</td>\n",
       "      <td>N</td>\n",
       "      <td>125.03</td>\n",
       "      <td>1.09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>PP_5335</td>\n",
       "      <td>N5-carboxyaminoimidazole ribonucleotide syntha...</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>214.31</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.03</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PP_5416</td>\n",
       "      <td>ATP synthase subunit delta (ATP synthase F(1) ...</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>152.72</td>\n",
       "      <td>0.02</td>\n",
       "      <td>1.17</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PP_5417</td>\n",
       "      <td>ATP synthase subunit b (ATP synthase F(0) sect...</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>202.53</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.84</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PP_5419</td>\n",
       "      <td>ATP synthase subunit a (ATP synthase F0 sector...</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>391.45</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PP_5420</td>\n",
       "      <td>ATP synthase protein I</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>91.60</td>\n",
       "      <td>0.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Locus number                                       Protein name  \\\n",
       "0        PP_0001      Probable chromosome-partitioning protein ParB   \n",
       "1        PP_0103        Cytochrome c oxidase subunit 2 (EC 7.1.1.9)   \n",
       "2        PP_0104        Cytochrome c oxidase subunit 1 (EC 7.1.1.9)   \n",
       "3        PP_0105         Cytochrome c oxidase assembly protein CtaG   \n",
       "4        PP_0106  cytochrome-c oxidase (EC 7.1.1.9) (Cytochrome ...   \n",
       "..           ...                                                ...   \n",
       "115      PP_5335  N5-carboxyaminoimidazole ribonucleotide syntha...   \n",
       "116      PP_5416  ATP synthase subunit delta (ATP synthase F(1) ...   \n",
       "117      PP_5417  ATP synthase subunit b (ATP synthase F(0) sect...   \n",
       "118      PP_5419  ATP synthase subunit a (ATP synthase F0 sector...   \n",
       "119      PP_5420                             ATP synthase protein I   \n",
       "\n",
       "    Source (Flux-RETAP or Heuristic) RbTn-Seq Essential?  \\\n",
       "0                                 FR                   N   \n",
       "1                                 FR                   N   \n",
       "2                                 FR                   N   \n",
       "3                                 FR                   N   \n",
       "4                                 FR                   N   \n",
       "..                               ...                 ...   \n",
       "115                                H                   Y   \n",
       "116                                H                   Y   \n",
       "117                                H                   Y   \n",
       "118                                H                   Y   \n",
       "119                                H                   Y   \n",
       "\n",
       "     Mean isoprenol titer (mg/L)  dCas9/control  Target/control  \\\n",
       "0                          58.03           1.04            0.08   \n",
       "1                         183.52           0.20             NaN   \n",
       "2                         130.41           1.11             NaN   \n",
       "3                         144.84           0.96             NaN   \n",
       "4                         125.03           1.09             NaN   \n",
       "..                           ...            ...             ...   \n",
       "115                       214.31           0.03            1.03   \n",
       "116                       152.72           0.02            1.17   \n",
       "117                       202.53           0.15            0.84   \n",
       "118                       391.45           0.02            0.02   \n",
       "119                        91.60           0.03             NaN   \n",
       "\n",
       "     Passed DBTL0 filter?  Used after DBTL2/3?  \n",
       "0                   False                False  \n",
       "1                   False                False  \n",
       "2                   False                False  \n",
       "3                   False                False  \n",
       "4                   False                False  \n",
       "..                    ...                  ...  \n",
       "115                 False                False  \n",
       "116                 False                False  \n",
       "117                  True                False  \n",
       "118                 False                False  \n",
       "119                 False                False  \n",
       "\n",
       "[120 rows x 9 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f6197b13-9bf3-41b6-bee0-5fa542c9a748",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Locus number</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>Source (Flux-RETAP or Heuristic)</th>\n",
       "      <th>RbTn-Seq Essential?</th>\n",
       "      <th>Mean isoprenol titer (mg/L)</th>\n",
       "      <th>dCas9/control</th>\n",
       "      <th>Target/control</th>\n",
       "      <th>Passed DBTL0 filter?</th>\n",
       "      <th>Used after DBTL2/3?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PP_0814</td>\n",
       "      <td>Cytochrome bo(3) ubiquinol oxidase subunit 3 (...</td>\n",
       "      <td>FR</td>\n",
       "      <td>Y</td>\n",
       "      <td>312.4</td>\n",
       "      <td>0.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Locus number                                       Protein name  \\\n",
       "25      PP_0814  Cytochrome bo(3) ubiquinol oxidase subunit 3 (...   \n",
       "\n",
       "   Source (Flux-RETAP or Heuristic) RbTn-Seq Essential?  \\\n",
       "25                               FR                   Y   \n",
       "\n",
       "    Mean isoprenol titer (mg/L)  dCas9/control  Target/control  \\\n",
       "25                        312.4           0.03             NaN   \n",
       "\n",
       "    Passed DBTL0 filter?  Used after DBTL2/3?  \n",
       "25                  True                 True  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_table[df_table['Locus number'] == 'PP_0814']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "45a70aec-3338-4c16-9fd4-ceff70250b6f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orgId</th>\n",
       "      <th>locusId</th>\n",
       "      <th>sysName</th>\n",
       "      <th>geneName</th>\n",
       "      <th>desc</th>\n",
       "      <th>set10IT006 L-Arginine (C)</th>\n",
       "      <th>set10IT008 L-Glutamic (C)</th>\n",
       "      <th>set10IT010 L-Histidine (C)</th>\n",
       "      <th>set10IT017 D-Glucose (C) with 20AA_mix_minus_Ala</th>\n",
       "      <th>set10IT018 D-Glucose (C) with 20AA_mix_minus_Arg</th>\n",
       "      <th>...</th>\n",
       "      <th>set8IT089 Adaptation in M9 media timecourse 4; day 2</th>\n",
       "      <th>set8IT090 Adaptation in M9 media timecourse 4; day 3</th>\n",
       "      <th>set8IT092 24hr timepoint from 1L M9/1% dextrose BATCH fermentation in 2L Sartorius Biostat reactor kept at 30% DO.</th>\n",
       "      <th>set8IT093 48hr timepoint from 1L M9/1% dextrose BATCH fermentation in 2L Sartorius Biostat reactor kept at 10% DO.</th>\n",
       "      <th>set8IT094 48hr timepoint from 1L M9/1% dextrose BATCH fermentation in 2L Sartorius Biostat reactor kept at 30% DO.</th>\n",
       "      <th>set8IT095 72hr timepoint from 1L M9/1% dextrose BATCH fermentation in 2L Sartorius Biostat reactor kept at 10% DO.</th>\n",
       "      <th>set8IT096 72hr timepoint from 1L M9/1% dextrose BATCH fermentation in 2L Sartorius Biostat reactor kept at 30% DO.</th>\n",
       "      <th>set9IT020 24hr timepoint from 1L M9/1% dextrose BATCH fermentation in 2L Sartorius Biostat reactor kept at 30% DO with 20% v/v dodecane overlay</th>\n",
       "      <th>set9IT022 48hr timepoint from 1L M9/1% dextrose BATCH fermentation in 2L Sartorius Biostat reactor kept at 30% DO with 20% v/v dodecane overlay</th>\n",
       "      <th>set9IT025 72hr timepoint from 1L M9/1% dextrose BATCH fermentation in 2L Sartorius Biostat reactor kept at 30% DO with 20% v/v oleyl alcohol overlay</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 305 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [orgId, locusId, sysName, geneName, desc, set10IT006 L-Arginine (C), set10IT008 L-Glutamic (C), set10IT010 L-Histidine (C), set10IT017 D-Glucose (C) with 20AA_mix_minus_Ala, set10IT018 D-Glucose (C) with 20AA_mix_minus_Arg, set10IT019 D-Glucose (C) with 20AA_mix_minus_Asn, set10IT020 D-Glucose (C) with 20AA_mix_minus_Asp, set10IT021 D-Glucose (C) with 20AA_mix_minus_Cys, set10IT022 D-Glucose (C) with 20AA_mix_minus_Glu, set10IT023 D-Glucose (C) with 20AA_mix_minus_Gln, set10IT024 D-Glucose (C) with 20AA_mix_minus_Gly, set10IT025 D-Glucose (C) with 20AA_mix_minus_His, set10IT026 D-Glucose (C) with 20AA_mix_minus_Ile, set10IT027 D-Glucose (C) with 20AA_mix_minus_Leu, set10IT028 D-Glucose (C) with 20AA_mix_minus_Lys, set10IT029 D-Glucose (C) with 20AA_mix_minus_Met, set10IT031 D-Glucose (C) with 20AA_mix_minus_Pro, set10IT032 D-Glucose (C) with 20AA_mix_minus_Ser, set10IT033 D-Glucose (C) with 20AA_mix_minus_Thr, set10IT034 D-Glucose (C) with 20AA_mix_minus_Trp, set10IT035 D-Glucose (C) with 20AA_mix_minus_Tyr, set10IT036 D-Glucose (C) with 20AA_mix_minus_Val, set10IT037 D-Glucose (C) with 20AA_mix, set10IT043 succinate (C), set10IT044 a-Ketoglutaric (C), set10IT053 butyrate (C), set12IT001 Vanillin (C), set12IT002 1,2-Propanediol (C), set12IT004 propionate (C), set12IT007 2-methyl-1-butanol (C), set12IT008 1,5-Pentanediol (C), set12IT009 1,4-Butanediol (C), set12IT014 Vanillic Acid (C), set12IT015 Protocatechuic Acid (C), set12IT016 4-Hydroxybenzoic Acid (C), set12IT017 Protocatechuic Acid (C), set12IT023 L-Lysine (C), set12IT024 p-Coumaric (C), set12IT025 benzoic (C), set12IT026 Ferulic Acid (C), set12IT027 D-Glucose (C), set12IT028 L-Glutamic (C), set12IT029 Phenylacetic (C), set12IT031 Vanillin (C), set12IT032 1,2-Propanediol (C), set12IT034 propionate (C), set12IT037 2-methyl-1-butanol (C), set12IT038 1,5-Pentanediol (C), set12IT039 1,4-Butanediol (C), set12IT044 Vanillic Acid (C), set12IT046 4-Hydroxybenzoic Acid (C), set12IT053 L-Lysine (C), set12IT054 p-Coumaric (C), set12IT055 benzoic (C), set12IT056 Ferulic Acid (C), set12IT057 D-Glucose (C), set12IT058 L-Glutamic (C), set12IT059 Phenylacetic (C), set15IT054 Butyl stearate (C), set15IT057 Hexanoic (C), set15IT058 Decanoic (C), set15IT059 Lauric (C), set15IT060 Valeric (C), set15IT061 Octanoic (C), set15IT062 Nonanoic (C), set15IT063 Tween 20 (C), set15IT064 Myristic (C), set15IT065 butyrate (C), set15IT066 Oleic (C), set15IT067 Heptanoic (C), set15IT068 Butyl stearate (C), set15IT071 Hexanoic (C), set15IT072 Decanoic (C), set15IT073 Lauric (C), set15IT074 Valeric (C), set15IT075 Octanoic (C), set15IT076 Nonanoic (C), set15IT077 Tween 20 (C), set15IT078 Myristic (C), set15IT079 butyrate (C), set15IT080 Oleic (C), set15IT081 Heptanoic (C), set16IT004 Butanol (C), set16IT005 Butanol (C), set16IT006 1,3-Butandiol (C), set16IT007 1,3-Butandiol (C), set16IT008 3-methyl-3-butenol (C), set16IT009 3-methyl-3-butenol (C), set16IT010 3-methyl-1-butanol (C), set16IT011 3-methyl-1-butanol (C), set16IT012 1-Pentanol (C), set16IT013 1-Pentanol (C), set16IT014 Ethanol (C), set16IT015 Ethanol (C), set1IT078 Glucose (C) (40mM), ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 305 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rbtnseq_df_full[rbtnseq_df_full['locusId'] == 'PP_0814']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "c8c26f08-8c0f-4280-90a4-c3e1841ea9b6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'PP_0814' in rbtnseq_df_full['locusId'].values"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
